{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CME Model of JCVI-syn3A Genetic Information Processes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyLM import *\n",
    "from pyLM.units import *\n",
    "from pySTDLM import *\n",
    "from pySTDLM.PostProcessing import *\n",
    "import math as m\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "import os\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "import csv\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import importlib\n",
    "from collections import defaultdict, OrderedDict\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load in model data such as the genome and proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load all necessary files\n",
    "\n",
    "# The annotation matches MMSYN1* IDs with JCVISYN3* IDs (or \"locus tags\").\n",
    "annotatPD = pd.read_excel(\"../model_data/FBA/Syn3A_annotation_compilation.xlsx\",\n",
    "                         sheet_name=\"Syn3A_annotation_compilation_condensed\")\n",
    "\n",
    "# The genome data matches \"locus tags\" with AOE* protein IDs.\n",
    "# It provides both the gene sequence, needed for transcription reactions in the ODE model,\n",
    "# and the protein sequence, needed for translation reactions in the model.\n",
    "# This is the NCBI Gene Bank-formated file (https://www.ncbi.nlm.nih.gov/nuccore/CP014992.1).\n",
    "\n",
    "genomeFile2 = '../model_data/syn2.gb'\n",
    "genome2 = next(SeqIO.parse(genomeFile2, \"gb\"))\n",
    "\n",
    "# This is the NCBI Gene Bank-formated file (https://www.ncbi.nlm.nih.gov/nuccore/CP016816.2).\n",
    "genomeFile3A = '../model_data/syn3A.gb'\n",
    "genome3A = next(SeqIO.parse(genomeFile3A, \"gb\"))\n",
    "\n",
    "# The proteomics matches AOE IDs with quantitative proteomics data.\n",
    "proteomPD = pd.read_excel(\"../model_data/proteomics.xlsx\", sheet_name=\"Proteomics\", skiprows=[0] )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Definitions to Get Protein Counts and Gene Sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create list of proteins with no proteomics data\n",
    "ptnNoQuant = set()\n",
    "\n",
    "def getPtnCount(newMetID, jcvi2ID):\n",
    "    \n",
    "    # Check if protein quantification is available.\n",
    "    try:\n",
    "        if jcvi2ID.startswith(\"JCVIman_\"):\n",
    "            aoeID = manGPRPD.loc[ manGPRPD.MM == jcvi2ID.replace(\"JCVIman_\",\"\"), \"AOE\" ].values[0]\n",
    "        else:\n",
    "            aoeID = J2toAOE[ jcvi2ID ]\n",
    "        \n",
    "        ptnCount = max(defaultPtnCount,round(proteomPD.loc[ proteomPD.Protein == aoeID ].iloc[0,21]))\n",
    "#         \n",
    "        ptnName  = proteomPD.loc[ proteomPD.Protein == aoeID ].iloc[0,1].replace(\n",
    "            \" [synthetic bacterium JCVI-Syn3.0]\",\"\")\n",
    "        \n",
    "#         ptnConcentration = ptnCount*countToMiliMol\n",
    "    except:\n",
    "        print(\"WARNING: No protein count for\", newMetID)\n",
    "        print(\"Using default protein concentration.\")\n",
    "\n",
    "        ptnName = newMetID\n",
    "        ptnCount = defaultPtnCount\n",
    "#         ptnConcentration = defaultPtnConcentration\n",
    "\n",
    "        ptnNoQuant.add(newMetID)\n",
    "    \n",
    "    return ptnCount, ptnName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getSequences(jcvi3AID):\n",
    "    # returns genomic and protein sequences\n",
    "    try:\n",
    "        rnasequence = genomePtnLocDict[jcvi3AID].extract(genome3A.seq).transcribe()\n",
    "        \n",
    "        # Using translation table 11 from NCBI: \"Bacterial, Archaeal and Plant Plastid Code\"\n",
    "        # https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi#SG4\n",
    "        aasequence  = genomePtnLocDict[jcvi3AID].extract(genome3A.seq).transcribe().translate(table=4)\n",
    "        \n",
    "    except:\n",
    "        aasequence  = 0\n",
    "        rnasequence = 0\n",
    "    \n",
    "    return rnasequence, aasequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# getSequences('JCVISYN3A_0001')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getRNAsequences(jcvi3AID):\n",
    "    # returns genomic and protein sequences\n",
    "    try:\n",
    "        rnasequence = genomeRnaLocDict[jcvi3AID].extract(genome3A.seq).transcribe()\n",
    "        \n",
    "    except:\n",
    "        rnasequence = 0\n",
    "    \n",
    "    return rnasequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "AOEtoJ2 = dict()\n",
    "J2toAOE = dict()\n",
    "genomePtnLocDict = dict()\n",
    "genomeRnaLocDict = dict()\n",
    "Locus3A = []\n",
    "\n",
    "for f in genome2.features:\n",
    "    if f.type == \"CDS\":\n",
    "        JCVSYN2_tag = f.qualifiers['locus_tag'][0]\n",
    "        #print(JCVSYN2_tag)\n",
    "        # Not all entries have an AOE protein_id\n",
    "        if('protein_id' in f.qualifiers.keys()):\n",
    "            AOE_locus = f.qualifiers['protein_id'][0]\n",
    "            AOEtoJ2[AOE_locus] = JCVSYN2_tag\n",
    "            J2toAOE[JCVSYN2_tag] = AOE_locus\n",
    "#             genomeLocDict[JCVSYN2_tag] = f.location\n",
    "        else:\n",
    "            print(\"Locus \", JCVSYN2_tag, \" has no AOE id!\")\n",
    "    if f.type == \"rRNA\":\n",
    "        JCVSYN2_tag = f.qualifiers['locus_tag'][0]\n",
    "#         genomeLocDict[JCVSYN2_tag] = f.location\n",
    "    if f.type == \"tRNA\":\n",
    "        JCVSYN2_tag = f.qualifiers['locus_tag'][0]\n",
    "#         genomeLocDict[JCVSYN2_tag] = f.location\n",
    "        \n",
    "for f in genome3A.features:\n",
    "    if f.type == \"CDS\":\n",
    "        JCVSYN3A_tag = f.qualifiers['locus_tag'][0]\n",
    "        Locus3A.append(JCVSYN3A_tag)\n",
    "        #print(JCVSYN2_tag)\n",
    "        # Not all entries have an AOE protein_id\n",
    "        if('protein_id' in f.qualifiers.keys()):\n",
    "#             AOE_locus = f.qualifiers['protein_id'][0]\n",
    "#             AOEtoJ2[AOE_locus] = JCVSYN2_tag\n",
    "#             J2toAOE[JCVSYN2_tag] = AOE_locus\n",
    "            genomePtnLocDict[JCVSYN3A_tag] = f.location\n",
    "        else:\n",
    "            print(\"Locus \", JCVSYN3A_tag, \" is pseudo.\")\n",
    "    if f.type == \"rRNA\":\n",
    "        JCVSYN3A_tag = f.qualifiers['locus_tag'][0]\n",
    "        Locus3A.append(JCVSYN3A_tag)\n",
    "        genomeRnaLocDict[JCVSYN3A_tag] = f.location\n",
    "    if f.type == \"tRNA\":\n",
    "        JCVSYN3A_tag = f.qualifiers['locus_tag'][0]\n",
    "        Locus3A.append(JCVSYN3A_tag)\n",
    "        genomeRnaLocDict[JCVSYN3A_tag] = f.location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomePtnLocDict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialize the Simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create our simulation object\n",
    "sim=CME.CMESimulation(name=\"Min Cell Genetic Processes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Lists of Proteins for Metabolism and with Specific Naming in the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PtnMetDF = pd.read_csv(\"../model_data/protein_metabolites.csv\")\n",
    "PtnMetDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "riboPtnMetDF = pd.read_csv(\"../model_data/ribo_protein_metabolites.csv\")\n",
    "riboPtnMetDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RNApolPtnMetDF = pd.read_csv(\"../model_data/RNApol_proteins.csv\")\n",
    "RNApolPtnMetDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rrnaMetDF_1 = pd.read_csv(\"../model_data/rrna_metabolites_1.csv\")\n",
    "rrnaMetDF_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rrnaMetDF_2 = pd.read_csv(\"../model_data/rrna_metabolites_2.csv\")\n",
    "rrnaMetDF_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trnaMetDF = pd.read_csv(\"../model_data/trna_metabolites.csv\")\n",
    "trnaMetDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MetPtnGenes = ['MMSYN1_0445', 'MMSYN1_0220',\n",
    " 'MMSYN1_0131', 'MMSYN1_0727', 'MMSYN1_0607',\n",
    " 'MMSYN1_0451', 'MMSYN1_0606', 'MMSYN1_0729',\n",
    " 'MMSYN1_0213', 'MMSYN1_0221', 'MMSYN1_0495',\n",
    " 'MMSYN1_0494', 'MMSYN1_0493', 'MMSYN1_0726',\n",
    " 'MMSYN1_0435', 'MMSYN1_0475', 'MMSYN1_0227',\n",
    " 'MMSYN1_0228', 'MMSYN1_0229', 'MMSYN1_0230',\n",
    " 'MMSYN1_0316', 'MMSYN1_0262', 'MMSYN1_0800',\n",
    " 'MMSYN1_0831', 'MMSYN1_0733', 'MMSYN1_0732',\n",
    " 'MMSYN1_0432', 'MMSYN1_0381', 'MMSYN1_0687',\n",
    " 'MMSYN1_0688', 'MMSYN1_0689', 'MMSYN1_0012',\n",
    " 'MMSYN1_0519', 'MMSYN1_0260', 'MMSYN1_0634',\n",
    " 'MMSYN1_0837', 'MMSYN1_0126', 'MMSYN1_0535',\n",
    " 'MMSYN1_0613', 'MMSYN1_0308', 'MMSYN1_0061',\n",
    " 'MMSYN1_0222', 'MMSYN1_0282', 'MMSYN1_0287',\n",
    " 'MMSYN1_0076', 'MMSYN1_0064', 'MMSYN1_0288',\n",
    " 'MMSYN1_0528', 'MMSYN1_0529', 'MMSYN1_0163',\n",
    " 'MMSYN1_0405', 'MMSYN1_0614', 'MMSYN1_0380',\n",
    " 'MMSYN1_0378', 'MMSYN1_0259', 'MMSYN1_0291',\n",
    " 'MMSYN1_0823', 'MMSYN1_0684', 'MMSYN1_0390',\n",
    " 'MMSYN1_0799', 'MMSYN1_0443', 'MMSYN1_0413',\n",
    " 'MMSYN1_0747', 'MMSYN1_0651', 'MMSYN1_0771',\n",
    " 'MMSYN1_0772', 'MMSYN1_0773', 'MMSYN1_0819',\n",
    " 'MMSYN1_0344', 'MMSYN1_0330', 'MMSYN1_0382',\n",
    " 'MMSYN1_0216', 'MMSYN1_0203', 'MMSYN1_0798',\n",
    " 'MMSYN1_0537', 'MMSYN1_0347', 'MMSYN1_0140',\n",
    " 'MMSYN1_0045', 'MMSYN1_0129', 'MMSYN1_0515',\n",
    " 'MMSYN1_0447', 'MMSYN1_0218', 'MMSYN1_0513',\n",
    " 'MMSYN1_0139', 'MMSYN1_0420', 'MMSYN1_0616',\n",
    " 'MMSYN1_0617', 'MMSYN1_0419', 'MMSYN1_0117',\n",
    " 'MMSYN1_0512', 'MMSYN1_0304', 'MMSYN1_0875',\n",
    " 'MMSYN1_0214', 'MMSYN1_0147', 'MMSYN1_0115',\n",
    " 'MMSYN1_0813', 'MMSYN1_0814', 'MMSYN1_0114',\n",
    " 'MMSYN1_0697', 'MMSYN1_0113']\n",
    "# MetPtnGenes\n",
    "\n",
    "MetLocusNums = []\n",
    "\n",
    "for gene in MetPtnGenes:\n",
    "    locusNum = gene.split('_')[1]\n",
    "    MetLocusNums.append(locusNum)\n",
    "    \n",
    "# MetLocusNums"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "named_PTN_list = []\n",
    "\n",
    "for index, row in riboPtnMetDF.iterrows():\n",
    "    named_PTN_list.append(row[\"gene\"]) \n",
    "\n",
    "    \n",
    "for index, row in PtnMetDF.iterrows():\n",
    "    named_PTN_list.append(row[\"gene\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Transcription and Translation Rate Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ModelSpecies = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "defaultPtnCount = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseMap = OrderedDict({ \"A\":\"M_atp_c\", \"U\":\"M_utp_c\", \"G\":\"M_gtp_c\", \"C\":\"M_ctp_c\" })\n",
    "# baseMapToMonoP = OrderedDict({ \"A\":\"M_amp_c\", \"U\":\"M_ump_c\", \"G\":\"M_gmp_c\", \"C\":\"M_cmp_c\" })\n",
    "\n",
    "# Global parameters for transcription\n",
    "rnaPolKcat = 0.155*187/493*25 # nt/s\n",
    "rnaPolK0 = 1e-4 #mM\n",
    "rnaPolKd = 0.01 #mM\n",
    "\n",
    "rrnaPolKcat = 90*2 # nt/s\n",
    "\n",
    "krnadeg = 0.00578/2 # 1/s\n",
    "ptnDegRate = 7.70e-06 # 1/s\n",
    "\n",
    "ATPconc = 1.04 #mM\n",
    "UTPconc = 0.68 #mM\n",
    "CTPconc = 0.34 #mM\n",
    "GTPconc = 0.68 #mM\n",
    "\n",
    "# Cell radius (meters):\n",
    "# r_cell = 2.5*(10**-7)\n",
    "r_cell = 2.0*(10**-7) # m\n",
    "\n",
    "CytoVolume = (4*np.pi/3)*1000*r_cell**3 # L\n",
    "cellVolume = CytoVolume\n",
    "\n",
    "# print(cellVolume)\n",
    "\n",
    "# Avogadro:\n",
    "avgdr   = 6.022e23 # molec/mol\n",
    "Avognum = avgdr\n",
    "\n",
    "countToMiliMol = 1000/(avgdr*cellVolume)\n",
    "\n",
    "RnaPconc = 187*countToMiliMol # mM\n",
    "\n",
    "# Global parameter for degradation of mRNAs\n",
    "rnaDegRate = 0.00578/2 # 1/s\n",
    "\n",
    "# Create a map for rna sequence to NTP concentration.\n",
    "baseMap = OrderedDict({ \"A\":ATPconc, \"U\":UTPconc, \"G\":GTPconc, \"C\":CTPconc })\n",
    "\n",
    "# Create Dictionaries to map tRNAs to associated aa abbreviations in protein sequences.\n",
    "aaMap = OrderedDict({\"A\":\"M_ala__L_c\", \"R\":\"M_arg__L_c\", \n",
    "    \"N\":\"M_asn__L_c\", \"D\":\"M_asp__L_c\", \"C\":\"M_cys__L_c\", \"E\":\"M_glu__L_c\", \"Q\":\"M_gln__L_c\", \"G\":\"M_gly_c\", \n",
    "    \"H\":\"M_his__L_c\", \"I\":\"M_ile__L_c\", \"L\":\"M_leu__L_c\", \"K\":\"M_lys__L_c\", \"M\":\"M_met__L_c\", \"F\":\"M_phe__L_c\", \n",
    "    \"P\":\"M_pro__L_c\", \"S\":\"M_ser__L_c\", \"T\":\"M_thr__L_c\", \"W\":\"M_trp__L_c\", \"Y\":\"M_tyr__L_c\", \"V\":\"M_val__L_c\",\n",
    "    \"*\":\"Stop_Codon\"})\n",
    "\n",
    "aaTRNAMap = OrderedDict({\"A\":\"M_alatrna_c\", \"R\":\"M_argtrna_c\", \n",
    "    \"N\":\"M_asntrna_c\", \"D\":\"M_asptrna_c\", \"C\":\"M_cystrna_c\", \"E\":\"M_glutrna_c\", \"Q\":\"M_glntrna_c\", \"G\":\"M_glytrna_c\", \n",
    "    \"H\":\"M_histrna_c\", \"I\":\"M_iletrna_c\", \"L\":\"M_leutrna_c\", \"K\":\"M_lystrna_c\", \"M\":\"M_mettrna_c\", \"F\":\"M_phetrna_c\", \n",
    "    \"P\":\"M_protrna_c\", \"S\":\"M_sertrna_c\", \"T\":\"M_thrtrna_c\", \"W\":\"M_trptrna_c\", \"Y\":\"M_tyrtrna_c\", \"V\":\"M_valtrna_c\"})\n",
    "\n",
    "aaTRNAFreeMap = OrderedDict({\"A\":\"M_trnaala_c\", \"R\":\"M_trnaarg_c\", \n",
    "    \"N\":\"M_trnaasn_c\", \"D\":\"M_trnaasp_c\", \"C\":\"M_trnacys_c\", \"E\":\"M_trnaglu_c\", \"Q\":\"M_trnagln_c\", \"G\":\"M_trnagly_c\", \n",
    "    \"H\":\"M_trnahis_c\", \"I\":\"M_trnaile_c\", \"L\":\"M_trnaleu_c\", \"K\":\"M_trnalys_c\", \"M\":\"M_trnamet_c\", \"F\":\"M_trnaphe_c\", \n",
    "    \"P\":\"M_trnapro_c\", \"S\":\"M_trnaser_c\", \"T\":\"M_trnathr_c\", \"W\":\"M_trnatrp_c\", \"Y\":\"M_trnatyr_c\", \"V\":\"M_trnaval_c\"})\n",
    "\n",
    "\n",
    "# Global parameters for translation\n",
    "riboKcat = 5 # 1/s\n",
    "riboK0 = 4*25e-6 # \n",
    "riboKd = 0.0001 # \n",
    "\n",
    "ribosomeConc = 500*countToMiliMol # mM\n",
    "\n",
    "# Concentration of charged tRNA\n",
    "ctRNAconc = 150*countToMiliMol # mM\n",
    "\n",
    "# Global parameter for degradation of proteins\n",
    "# Derived from eLife's model, using average protein half life of 25 hours. \n",
    "ptnDegRate = 7.70e-06 # 1/s\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define how to calculate transcription rate constants as in equation 3 for transcription reactions.\n",
    "# Uses mature transcript length and proteomics for promoter strength.\n",
    "\n",
    "def TranscriptRate(rnaMetID, ptnMetID, rnasequence, jcvi2ID):\n",
    "    # Add trascription reaction\n",
    "    \n",
    "    # Check that we know all bases used in the sequence\n",
    "    if ( set(rnasequence) - set(baseMap.keys()) ):\n",
    "        raise Exception(\"Unknown base(s) in RNA sequence {}\".format(set(rnasequence) - set(baseMap.keys())) )\n",
    "    \n",
    "    # Count how many times each base is used\n",
    "    baseCount = defaultdict(int)\n",
    "    for base in set(rnasequence):\n",
    "        baseCount[base] = rnasequence.count(base)\n",
    "    \n",
    "\n",
    "    ptnCount, ptnName = getPtnCount(ptnMetID, jcvi2ID)\n",
    "    \n",
    "    kcat_mod = min(rnaPolKcat*(ptnCount/(180)),2*90)\n",
    "\n",
    "    \n",
    "    # Add total number of monomers to parameter dict\n",
    "    \n",
    "    CMono1 = baseMap[ rnasequence[0] ]\n",
    "    \n",
    "    CMono2 = baseMap[ rnasequence[1] ]\n",
    "\n",
    "    n_tot = sum(list(baseCount.values()))\n",
    "\n",
    "    NMono_A = baseCount[\"A\"]\n",
    "    \n",
    "    NMono_U = baseCount[\"C\"]\n",
    "    \n",
    "    NMono_C = baseCount[\"G\"]\n",
    "    \n",
    "    NMono_G = baseCount[\"U\"]\n",
    "    \n",
    "    NMonoDict = [NMono_A,NMono_C,NMono_G,NMono_U]\n",
    "    print(NMonoDict)\n",
    "    \n",
    "    \n",
    "    NMonoSum = NMono_A*rnaPolKd/ATPconc + NMono_C*rnaPolKd/CTPconc + NMono_U*rnaPolKd/UTPconc + NMono_G*rnaPolKd/GTPconc\n",
    "    \n",
    "\n",
    "    k_transcription = kcat_mod / ((1+rnaPolK0/RnaPconc)*(rnaPolKd**2)/(CMono1*CMono2) + NMonoSum + n_tot - 1)\n",
    "    \n",
    "    return k_transcription\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define transcription rate for ribosomal protein-coding genes.\n",
    "\n",
    "def riboTranscriptRate(rnaMetID, ptnMetID, rnasequence, jcvi2ID):\n",
    "    # Add trascription reaction\n",
    "    \n",
    "    # Check that we know all bases used in the sequence\n",
    "    if ( set(rnasequence) - set(baseMap.keys()) ):\n",
    "        raise Exception(\"Unknown base(s) in RNA sequence {}\".format(set(rnasequence) - set(baseMap.keys())) )\n",
    "    \n",
    "    # Count how many times each base is used\n",
    "    baseCount = defaultdict(int)\n",
    "    for base in set(rnasequence):\n",
    "        baseCount[base] = rnasequence.count(base)\n",
    "        \n",
    "    ptnCount, ptnName = getPtnCount(ptnMetID, jcvi2ID)\n",
    "    \n",
    "#     kcat_mod = min(rnaPolKcat*(ptnCount/(180)),2*85)\n",
    "    \n",
    "    kcat_mod = rnaPolKcat#*(339/(180))\n",
    "    \n",
    "    # The rate form needs specific sequence data for the first two monomers:\n",
    "#     paramDict[\"CMonoA\"] = baseMap[ rnasequence[0] ]\n",
    "#     paramDict[\"CMonoB\"] = baseMap[ rnasequence[1] ]\n",
    "#     paramDict[\"KDA\"] = rnaPolKd # Since we are current;y using the same Kd for all nucleotides\n",
    "#     paramDict[\"KDB\"] = rnaPolKd \n",
    "    \n",
    "    # Add total number of monomers to parameter dict\n",
    "\n",
    "    n_tot = sum(list(baseCount.values()))\n",
    "\n",
    "    NMono_A = baseCount[\"A\"]\n",
    "    \n",
    "    NMono_U = baseCount[\"C\"]\n",
    "    \n",
    "    NMono_C = baseCount[\"G\"]\n",
    "    \n",
    "    NMono_G = baseCount[\"U\"]\n",
    "    \n",
    "    NMonoDict = [NMono_A,NMono_C,NMono_G,NMono_U]\n",
    "    print(NMonoDict)\n",
    "    \n",
    "    CMono1 = baseMap[ rnasequence[0] ]\n",
    "    \n",
    "    CMono2 = baseMap[ rnasequence[1] ]\n",
    "    \n",
    "    NMonoSum = NMono_A*rnaPolKd/ATPconc + NMono_C*rnaPolKd/CTPconc + NMono_U*rnaPolKd/UTPconc + NMono_G*rnaPolKd/GTPconc\n",
    "\n",
    "    k_transcription = kcat_mod / ((1+rnaPolK0/RnaPconc)*(rnaPolKd**2)/(CMono1*CMono2) + NMonoSum + n_tot - 1)\n",
    "    \n",
    "    return k_transcription"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define transcription rate for tRNA genes.\n",
    "\n",
    "def trnaTranscriptRate(rnaMetID, rnasequence):\n",
    "    # Add trascription reaction\n",
    "    \n",
    "    # Check that we know all bases used in the sequence\n",
    "    if ( set(rnasequence) - set(baseMap.keys()) ):\n",
    "        raise Exception(\"Unknown base(s) in RNA sequence {}\".format(set(rnasequence) - set(baseMap.keys())) )\n",
    "    \n",
    "    # Count how many times each base is used\n",
    "    baseCount = defaultdict(int)\n",
    "    for base in set(rnasequence):\n",
    "        baseCount[base] = rnasequence.count(base)\n",
    "    \n",
    "    # The rate form needs specific sequence data for the first two monomers:\n",
    "#     paramDict[\"CMonoA\"] = baseMap[ rnasequence[0] ]\n",
    "#     paramDict[\"CMonoB\"] = baseMap[ rnasequence[1] ]\n",
    "#     paramDict[\"KDA\"] = rnaPolKd # Since we are current;y using the same Kd for all nucleotides\n",
    "#     paramDict[\"KDB\"] = rnaPolKd \n",
    "    \n",
    "    # Add total number of monomers to parameter dict\n",
    "\n",
    "    n_tot = sum(list(baseCount.values()))\n",
    "\n",
    "    NMono_A = baseCount[\"A\"]\n",
    "    \n",
    "    NMono_U = baseCount[\"C\"]\n",
    "    \n",
    "    NMono_C = baseCount[\"G\"]\n",
    "    \n",
    "    NMono_G = baseCount[\"U\"]\n",
    "    \n",
    "    NMonoDict = [NMono_A,NMono_C,NMono_G,NMono_U]\n",
    "    print(NMonoDict)\n",
    "    \n",
    "    CMono1 = baseMap[ rnasequence[0] ]\n",
    "    \n",
    "    CMono2 = baseMap[ rnasequence[1] ]\n",
    "    \n",
    "    NMonoSum = NMono_A*rnaPolKd/ATPconc + NMono_C*rnaPolKd/CTPconc + NMono_U*rnaPolKd/UTPconc + NMono_G*rnaPolKd/GTPconc\n",
    "    \n",
    "    k_transcription = rnaPolKcat/ ((1+rnaPolK0/RnaPconc)*(rnaPolKd**2)/(CMono1*CMono2) + NMonoSum + n_tot - 1)\n",
    "    \n",
    "    return k_transcription"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define transcription rate for tRNA genes.\n",
    "\n",
    "def rrnaTranscriptRate(rnasequence):\n",
    "    # Add trascription reaction\n",
    "    \n",
    "    # Check that we know all bases used in the sequence\n",
    "    if ( set(rnasequence) - set(baseMap.keys()) ):\n",
    "        raise Exception(\"Unknown base(s) in RNA sequence {}\".format(set(rnasequence) - set(baseMap.keys())) )\n",
    "    \n",
    "    # Count how many times each base is used\n",
    "    baseCount = defaultdict(int)\n",
    "    for base in set(rnasequence):\n",
    "        baseCount[base] = rnasequence.count(base)\n",
    "    \n",
    "    # The rate form needs specific sequence data for the first two monomers:\n",
    "#     paramDict[\"CMonoA\"] = baseMap[ rnasequence[0] ]\n",
    "#     paramDict[\"CMonoB\"] = baseMap[ rnasequence[1] ]\n",
    "#     paramDict[\"KDA\"] = rnaPolKd # Since we are current;y using the same Kd for all nucleotides\n",
    "#     paramDict[\"KDB\"] = rnaPolKd \n",
    "    \n",
    "    # Add total number of monomers to parameter dict\n",
    "#     paramDict[\"n\"] = sum(list(baseCount.values()))\n",
    "    n_tot = sum(list(baseCount.values()))\n",
    "    \n",
    "    NMono_A = baseCount[\"A\"]\n",
    "    \n",
    "    NMono_U = baseCount[\"C\"]\n",
    "    \n",
    "    NMono_C = baseCount[\"G\"]\n",
    "    \n",
    "    NMono_G = baseCount[\"U\"]\n",
    "    \n",
    "    NMonoDict = [NMono_A,NMono_C,NMono_G,NMono_U]\n",
    "    print(NMonoDict)\n",
    "    \n",
    "    CMono1 = baseMap[ rnasequence[0] ]\n",
    "    \n",
    "    CMono2 = baseMap[ rnasequence[1] ]\n",
    "    \n",
    "    NMonoSum = NMono_A*rnaPolKd/ATPconc + NMono_C*rnaPolKd/CTPconc + NMono_U*rnaPolKd/UTPconc + NMono_G*rnaPolKd/GTPconc\n",
    "    \n",
    "    k_transcription = rrnaPolKcat / ((1+rnaPolK0/RnaPconc)*(rnaPolKd**2)/(CMono1*CMono2) + NMonoSum + n_tot - 1)\n",
    "    \n",
    "    return k_transcription"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define how to calculate translation rate constants as in equation 3 for translation reactions.\n",
    "\n",
    "def TranslatRate(rnaMetID, ptnMetID, rnasequence, aasequence):\n",
    "    # Add translation reaction\n",
    "    \n",
    "    # Considers amino acids up to the first stop codon.\n",
    "#     aasequence = aasequence[0:aasequence.find(\"*\")]\n",
    "    \n",
    "    # Check that we know all residues used in the sequence\n",
    "    if ( set(aasequence) - set(aaMap.keys()) ):\n",
    "        raise Exception(\"Unknown residue(s) in Protein sequence {}\".format(set(aasequence) - set(aaMap.keys())) )\n",
    "    \n",
    "    # Count how many times each residue is used\n",
    "    aaCount = defaultdict(int)\n",
    "    for aa in set(aasequence):\n",
    "        aaCount[aa] = aasequence.count(aa)\n",
    "    \n",
    "    NMono_A = aaCount[\"A\"]\n",
    "    NMono_R = aaCount[\"R\"]\n",
    "    NMono_N = aaCount[\"N\"]\n",
    "    NMono_D = aaCount[\"D\"]\n",
    "    NMono_C = aaCount[\"C\"]\n",
    "    NMono_E = aaCount[\"E\"]\n",
    "    NMono_Q = aaCount[\"Q\"]\n",
    "    NMono_H = aaCount[\"H\"]\n",
    "    NMono_I = aaCount[\"I\"]\n",
    "    NMono_L = aaCount[\"L\"]\n",
    "    NMono_K = aaCount[\"K\"]\n",
    "    NMono_M = aaCount[\"M\"]\n",
    "    NMono_P = aaCount[\"P\"]\n",
    "    NMono_S = aaCount[\"S\"]\n",
    "    NMono_T = aaCount[\"T\"]\n",
    "    NMono_W = aaCount[\"W\"]\n",
    "    NMono_Y = aaCount[\"Y\"]\n",
    "    NMono_G = aaCount[\"G\"]\n",
    "    NMono_F = aaCount[\"F\"]\n",
    "    NMono_V = aaCount[\"V\"]\n",
    "    \n",
    "    NStop = aaCount[\"*\"]\n",
    "    \n",
    "    if NStop > 1:\n",
    "        print(\"EXTRA STOP CODON: MISTAKE IN TRANSLATION\")\n",
    "    \n",
    "    NMonoDict = [NMono_A,NMono_R,NMono_N,NMono_D,NMono_C,NMono_E,NMono_Q,NMono_H,\n",
    "                 NMono_I,NMono_L,NMono_K,NMono_M,NMono_P,NMono_S,NMono_T,NMono_W,\n",
    "                 NMono_Y,NMono_G,NMono_F,NMono_V]\n",
    "    \n",
    "    NMonoSum = 0\n",
    "    \n",
    "    for nmono in range(0,len(NMonoDict)):\n",
    "        NMonoSum = NMonoSum + NMonoDict[nmono]*riboKd/ctRNAconc\n",
    "        \n",
    "    n_tot = sum(list(aaCount.values()))\n",
    "\n",
    "    baseCount = defaultdict(int)\n",
    "    for base in set(rnasequence):\n",
    "        baseCount[base] = rnasequence.count(base)\n",
    "        \n",
    "    transcript_length = sum(list(baseCount.values()))\n",
    "    \n",
    "    print(transcript_length)\n",
    "    \n",
    "    ribo_num = max(1,int(transcript_length/300)) #max(1,int(transcript_length/300))\n",
    "    \n",
    "    print(ribo_num)\n",
    "    \n",
    "    kcat_mod = riboKcat*ribo_num #*0.4\n",
    "    \n",
    "    k_translation = kcat_mod / ((1+riboK0/ribosomeConc)*(riboKd**2)/(ctRNAconc**2) + NMonoSum + n_tot - 1)\n",
    "    \n",
    "    return k_translation\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Definitions to Add RNA, Proteins, and Reactions to the Simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def addPtn(jcvi3AID):\n",
    "    locusNum = jcvi3AID.split('_')[1]\n",
    "    mmcode = 'MMSYN1_' + locusNum\n",
    "    # Checks if a translation to JCVISYN2* code is available\n",
    "    try:\n",
    "        jcvi2ID = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 13].strip()\n",
    "    except:\n",
    "        jcvi2ID = \"JCVIunk_\" + mmcode\n",
    "\n",
    "    print(mmcode, jcvi2ID, jcvi3AID)\n",
    "    \n",
    "    genes_in_model.append(jcvi3AID)\n",
    "\n",
    "    ptnMetID = 'M_PTN_' + jcvi3AID + '_c'\n",
    "    # If the protein is not in the model, add it:\n",
    "\n",
    "    ModelSpecies.append(ptnMetID)\n",
    "\n",
    "    ptnCount, ptnName = getPtnCount(ptnMetID, jcvi2ID)\n",
    "\n",
    "    print(ptnMetID, ptnCount)\n",
    "\n",
    "    geneMetID = jcvi3AID + '_gene'\n",
    "    \n",
    "    ModelSpecies.append(geneMetID)\n",
    "\n",
    "    # Get nucleotide and amino acid sequences, if available\n",
    "    rnasequence, aasequence = getSequences(jcvi3AID)\n",
    "    print(rnasequence)\n",
    "    print(aasequence)\n",
    "\n",
    "    if (rnasequence != 0) and (aasequence != 0):\n",
    "\n",
    "        rnaMetID = \"M_RNA_\" + jcvi3AID + \"_c\"\n",
    "        rnaName = \"(mRNA) \" + ptnName\n",
    "\n",
    "        ModelSpecies.append(rnaMetID)\n",
    "    \n",
    "        species = []\n",
    "        species = [geneMetID, rnaMetID, ptnMetID]\n",
    "        print(species)\n",
    "        sim.defineSpecies(species)\n",
    "\n",
    "        sim.addParticles(species = geneMetID, count = 1)\n",
    "        sim.addParticles(species = ptnMetID,  count = int(ptnCount))\n",
    "        sim.addParticles(species = rnaMetID, count = 1)\n",
    "        \n",
    "        TrscProd = [geneMetID,rnaMetID]\n",
    "        mRNAdegProd = []\n",
    "    \n",
    "        for i in range(len(rnasequence)):\n",
    "            TrscProd.append('ATP_trsc')\n",
    "            mRNAdegProd.append('ATP_mRNAdeg')\n",
    "            \n",
    "        TranslatProd = [rnaMetID,ptnMetID]\n",
    "        ptnDegProd = []\n",
    "        \n",
    "        for i in range(len(aasequence)):\n",
    "            TranslatProd.append('ATP_translat')\n",
    "            TranslatProd.append('ATP_translat')\n",
    "            ptnDegProd.append('ATP_ptndeg')\n",
    "    \n",
    "        sim.addReaction(geneMetID, tuple(TrscProd), TranscriptRate(rnaMetID, ptnMetID, rnasequence, jcvi2ID))\n",
    "        sim.addReaction(rnaMetID, tuple(mRNAdegProd) ,krnadeg)\n",
    "        sim.addReaction(rnaMetID, tuple(TranslatProd), TranslatRate(rnaMetID, ptnMetID, rnasequence, aasequence))\n",
    "        \n",
    "        sim.addReaction(ptnMetID,tuple(ptnDegProd),ptnDegRate)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def addNamedPtn(newMetID, mmcode):\n",
    "    \n",
    "    locusNum = mmcode.split('_')[1]\n",
    "\n",
    "    jcvi3AID = 'JCVISYN3A_' + locusNum\n",
    "    \n",
    "    genes_in_model.append(jcvi3AID)\n",
    "\n",
    "    # Checks if a translation to JCVISYN2* code is available\n",
    "    try:\n",
    "        jcvi2ID = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 13].strip()\n",
    "    except:\n",
    "        # If not, check for a manual AOE connection:\n",
    "        if mmcode in manGPRPD.MM.values:\n",
    "            jcvi2ID = \"JCVIman_\" + mmcode\n",
    "        else:\n",
    "            # If not, set an \"unknown\" code\n",
    "            jcvi2ID = \"JCVIunk_\" + mmcode + \"_\" + str(unkIter)\n",
    "\n",
    "    print(mmcode, jcvi2ID, jcvi3AID)\n",
    "\n",
    "    ptnCount, ptnName = getPtnCount(newMetID, jcvi2ID)\n",
    "    \n",
    "    ModelSpecies.append(newMetID)\n",
    "\n",
    "    print(newMetID, ptnCount)\n",
    "\n",
    "\n",
    "    geneMetID = jcvi3AID + '_gene'\n",
    "\n",
    "    ModelSpecies.append(geneMetID)\n",
    "    \n",
    "    rnasequence, aasequence = getSequences(jcvi3AID)\n",
    "    print(rnasequence)\n",
    "    print(aasequence)\n",
    "\n",
    "    if (rnasequence != 0) and (aasequence != 0):\n",
    "\n",
    "        rnaMetID = \"M_RNA_\" + jcvi3AID + \"_c\"\n",
    "        rnaName = \"(mRNA) \" + ptnName\n",
    "        \n",
    "    ModelSpecies.append(rnaMetID)\n",
    "\n",
    "    species = []\n",
    "    species = [geneMetID, rnaMetID, newMetID]\n",
    "    print(species)\n",
    "    sim.defineSpecies(species)\n",
    "\n",
    "    sim.addParticles(species = geneMetID, count = 1)\n",
    "    sim.addParticles(species = newMetID,  count = int(ptnCount))\n",
    "    sim.addParticles(species = rnaMetID, count = 1)\n",
    "    \n",
    "    TrscProd = [geneMetID,rnaMetID]\n",
    "    mRNAdegProd = []\n",
    "    \n",
    "    for i in range(len(rnasequence)):\n",
    "        TrscProd.append('ATP_trsc')\n",
    "        mRNAdegProd.append('ATP_mRNAdeg')\n",
    "        \n",
    "    TranslatProd = [rnaMetID,newMetID]\n",
    "    ptnDegProd = []\n",
    "        \n",
    "    for i in range(len(aasequence)):\n",
    "        TranslatProd.append('ATP_translat')\n",
    "        TranslatProd.append('ATP_translat')\n",
    "        ptnDegProd.append('ATP_ptndeg')\n",
    "    \n",
    "    sim.addReaction(geneMetID, tuple(TrscProd), TranscriptRate(rnaMetID, newMetID, rnasequence, jcvi2ID))\n",
    "    sim.addReaction(rnaMetID, tuple(mRNAdegProd) ,krnadeg)\n",
    "    sim.addReaction(rnaMetID, tuple(TranslatProd), TranslatRate(rnaMetID, newMetID, rnasequence, aasequence))\n",
    "\n",
    "    sim.addReaction(newMetID,tuple(ptnDegProd),ptnDegRate)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RiboPtnNames = ['Ribosomal Protein']\n",
    "RiboPtnLens = ['Gene Length (nt)']\n",
    "RiboPtnTrscRates = ['Transcription Rate']\n",
    "RiboPtnTranslatRates = ['Translation Rate']\n",
    "\n",
    "def addRiboPtn(newMetID, mmcode):\n",
    "    \n",
    "    locusNum = mmcode.split('_')[1]\n",
    "\n",
    "    jcvi3AID = 'JCVISYN3A_' + locusNum\n",
    "\n",
    "    genes_in_model.append(jcvi3AID)\n",
    "\n",
    "    # Checks if a translation to JCVISYN2* code is available\n",
    "    try:\n",
    "        jcvi2ID = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 13].strip()\n",
    "    except:\n",
    "        # If not, check for a manual AOE connection:\n",
    "        if mmcode in manGPRPD.MM.values:\n",
    "            jcvi2ID = \"JCVIman_\" + mmcode\n",
    "        else:\n",
    "            # If not, set an \"unknown\" code\n",
    "            jcvi2ID = \"JCVIunk_\" + mmcode + \"_\" + str(unkIter)\n",
    "            unkIter += 1 \n",
    "\n",
    "    print(mmcode, jcvi2ID, jcvi3AID)\n",
    "\n",
    "    ptnCount, ptnName = getPtnCount(newMetID, jcvi2ID)\n",
    "\n",
    "    print(newMetID, ptnCount)\n",
    "    \n",
    "    ModelSpecies.append(newMetID)\n",
    "\n",
    "    geneMetID = jcvi3AID + '_gene'\n",
    "\n",
    "    ModelSpecies.append(geneMetID)\n",
    "    \n",
    "    rnasequence, aasequence = getSequences(jcvi3AID)\n",
    "    print(rnasequence)\n",
    "    print(aasequence)\n",
    "\n",
    "    if (rnasequence != 0) and (aasequence != 0):\n",
    "\n",
    "        rnaMetID = \"M_RNA_\" + jcvi3AID + \"_c\"\n",
    "        rnaName = \"(mRNA) \" + ptnName\n",
    "        \n",
    "    ModelSpecies.append(rnaMetID)\n",
    "\n",
    "    species = []\n",
    "    species = [geneMetID, rnaMetID, newMetID]\n",
    "    print(species)\n",
    "    sim.defineSpecies(species)\n",
    "\n",
    "    sim.addParticles(species = geneMetID, count = 1)\n",
    "    sim.addParticles(species = newMetID,  count = int(ptnCount))\n",
    "    sim.addParticles(species = rnaMetID, count = 1)\n",
    "    \n",
    "    RiboPtnNames.append(newMetID)\n",
    "    \n",
    "    trsc_rate = riboTranscriptRate(rnaMetID, newMetID, rnasequence, jcvi2ID)\n",
    "    RiboPtnTrscRates.append(trsc_rate)\n",
    "    \n",
    "    translat_rate = TranslatRate(rnaMetID, newMetID, rnasequence, aasequence)\n",
    "    RiboPtnTranslatRates.append(translat_rate)\n",
    "    \n",
    "    rna_length = len(rnasequence)\n",
    "    RiboPtnLens.append(rna_length)\n",
    "    \n",
    "    TrscProd = [geneMetID,rnaMetID]\n",
    "    mRNAdegProd = []\n",
    "    \n",
    "    for i in range(len(rnasequence)):\n",
    "        TrscProd.append('ATP_trsc')\n",
    "        mRNAdegProd.append('ATP_mRNAdeg')\n",
    "        \n",
    "    TranslatProd = [rnaMetID,newMetID]\n",
    "        \n",
    "    for i in range(len(aasequence)):\n",
    "        TranslatProd.append('ATP_translat')\n",
    "        TranslatProd.append('ATP_translat')\n",
    "    \n",
    "    sim.addReaction(geneMetID, tuple(TrscProd), riboTranscriptRate(rnaMetID, newMetID, rnasequence, jcvi2ID))\n",
    "    sim.addReaction(rnaMetID, tuple(mRNAdegProd) ,krnadeg)\n",
    "    sim.addReaction(rnaMetID, tuple(TranslatProd), TranslatRate(rnaMetID, newMetID, rnasequence, aasequence))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tRNAadded = []\n",
    "\n",
    "def addtRNA(newMetID, mmcode):\n",
    "    \n",
    "    locusNum = mmcode.split('_')[1]\n",
    "\n",
    "    jcvi3AID = 'JCVISYN3A_' + locusNum\n",
    "\n",
    "    # Checks if a translation to JCVISYN2* code is available\n",
    "    try:\n",
    "        jcvi2ID = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 13].strip()\n",
    "    except:\n",
    "        # If not, check for a manual AOE connection:\n",
    "        if mmcode in manGPRPD.MM.values:\n",
    "            jcvi2ID = \"JCVIman_\" + mmcode\n",
    "        else:\n",
    "            # If not, set an \"unknown\" code\n",
    "            jcvi2ID = \"JCVIunk_\" + mmcode + \"_\" + str(unkIter)\n",
    "            unkIter += 1 \n",
    "\n",
    "    print(jcvi3AID)\n",
    "\n",
    "    geneMetID = jcvi3AID + '_gene'\n",
    "    \n",
    "    ModelSpecies.append(geneMetID)\n",
    "\n",
    "    rnasequence = getRNAsequences(jcvi3AID)\n",
    "    print(rnasequence)\n",
    "\n",
    "    gene = [geneMetID]\n",
    "    print(gene)\n",
    "    sim.defineSpecies(gene)\n",
    "\n",
    "    sim.addParticles(species = geneMetID, count = 1)\n",
    "\n",
    "    \n",
    "    if newMetID not in tRNAadded:\n",
    "        ModelSpecies.append(newMetID)\n",
    "        tRNA = [newMetID]\n",
    "        sim.defineSpecies(tRNA)\n",
    "        sim.addParticles(species = newMetID,  count = 38)\n",
    "        tRNAadded.append(newMetID)\n",
    "        \n",
    "    Products = [geneMetID,newMetID]\n",
    "        \n",
    "    for i in range(len(rnasequence)):\n",
    "        Products.append('ATP_trsc')\n",
    "        \n",
    "    \n",
    "    sim.addReaction(geneMetID, tuple(Products), trnaTranscriptRate(newMetID, rnasequence))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def addrRNA():\n",
    "    \n",
    "    rrna_gene_locs_1 = []\n",
    "    rrna_gene_locs_2 = []\n",
    "\n",
    "    rRNA_species = []\n",
    "    \n",
    "    for index, row in rrnaMetDF_1.iterrows():\n",
    "        newMetID = row[\"species\"]\n",
    "        jcvi3AID = row[\"gene\"]\n",
    "        \n",
    "#         rnasequence = getRNAsequences(jcvi3AID)\n",
    "        gene_str = str(genomeRnaLocDict[jcvi3AID])\n",
    "        start_nt = int(gene_str.split(':')[0].replace('[',''))\n",
    "        print(start_nt)\n",
    "        end_nt = int(gene_str.split(':')[1].replace('](-)',''))\n",
    "        print(end_nt)\n",
    "        rrna_gene_locs_1.append(start_nt)\n",
    "        rrna_gene_locs_1.append(end_nt)\n",
    "            \n",
    "        rRNA_species.append(newMetID)\n",
    "        \n",
    "    rrna_pos_1 = Seq(str(genome3A.seq[min(rrna_gene_locs_1):max(rrna_gene_locs_1)]))\n",
    "    \n",
    "    rrna_operon_1 = rrna_pos_1.reverse_complement()\n",
    "            \n",
    "    print(rrna_operon_1)\n",
    "    \n",
    "    \n",
    "    for index, row in rrnaMetDF_2.iterrows():\n",
    "        newMetID = row[\"species\"]\n",
    "        jcvi3AID = row[\"gene\"]\n",
    "        \n",
    "#         rnasequence = getRNAsequences(jcvi3AID)\n",
    "        gene_str = str(genomeRnaLocDict[jcvi3AID])\n",
    "        start_nt = int(gene_str.split(':')[0].replace('[',''))\n",
    "        print(start_nt)\n",
    "        end_nt = int(gene_str.split(':')[1].replace('](-)',''))\n",
    "        print(end_nt)\n",
    "        rrna_gene_locs_2.append(start_nt)\n",
    "        rrna_gene_locs_2.append(end_nt)\n",
    "        \n",
    "    rrna_pos_2 = Seq(str(genome3A.seq[min(rrna_gene_locs_2):max(rrna_gene_locs_2)]))\n",
    "    \n",
    "    rrna_operon_2 = rrna_pos_2.reverse_complement()\n",
    "            \n",
    "    print(rrna_operon_2)\n",
    "    \n",
    "    print(rRNA_species)\n",
    "    \n",
    "    sim.defineSpecies(rRNA_species)\n",
    "    \n",
    "    \n",
    "    for rRNA in rRNA_species:\n",
    "        \n",
    "        sim.addParticles(species = rRNA,  count = 1)\n",
    "            \n",
    "        ModelSpecies.append(rRNA)\n",
    "    \n",
    "    \n",
    "    \n",
    "    for index, row in rrnaMetDF_1.iterrows():\n",
    "        newMetID = row[\"species\"]\n",
    "        jcvi3AID = row[\"gene\"]\n",
    "        \n",
    "        geneMetID = jcvi3AID + '_gene'\n",
    "    \n",
    "        ModelSpecies.append(geneMetID)\n",
    "        \n",
    "        gene = [geneMetID]\n",
    "        print(gene)\n",
    "        sim.defineSpecies(gene)\n",
    "\n",
    "        sim.addParticles(species = geneMetID, count = 1)\n",
    "\n",
    "        print(jcvi3AID)\n",
    "\n",
    "        rnasequence = rrna_operon_1.transcribe()\n",
    "        \n",
    "        TrscProd = [geneMetID,newMetID]\n",
    "        \n",
    "        for i in range(int(len(rnasequence)/3)):\n",
    "            TrscProd.append('ATP_trsc')\n",
    "            \n",
    "        sim.addReaction(geneMetID, tuple(TrscProd), rrnaTranscriptRate(rnasequence))\n",
    "        \n",
    "        \n",
    "    for index, row in rrnaMetDF_2.iterrows():\n",
    "        newMetID = row[\"species\"]\n",
    "        jcvi3AID = row[\"gene\"]\n",
    "        \n",
    "        geneMetID = jcvi3AID + '_gene'\n",
    "    \n",
    "        ModelSpecies.append(geneMetID)\n",
    "        \n",
    "        gene = [geneMetID]\n",
    "        print(gene)\n",
    "        sim.defineSpecies(gene)\n",
    "\n",
    "        sim.addParticles(species = geneMetID, count = 1)\n",
    "\n",
    "        print(jcvi3AID)\n",
    "\n",
    "        rnasequence = rrna_operon_2.transcribe()\n",
    "        \n",
    "        TrscProd = [geneMetID,newMetID]\n",
    "        \n",
    "        for i in range(int(len(rnasequence)/3)):\n",
    "            TrscProd.append('ATP_trsc')\n",
    "        \n",
    "        sim.addReaction(geneMetID, tuple(TrscProd), rrnaTranscriptRate(rnasequence))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# REPLICATION INITIATION WRITTEN BY COLE CROTTY AND MODIFIED BY ZANE THORNBURG\n",
    "NA = 6.022*(10**(23))\n",
    "\n",
    "def addRepInit():\n",
    "    k_high = 7800*1000/NA/cellVolume\n",
    "    k_low = 35*1000/NA/cellVolume\n",
    "    k_on = 100*1000/NA/cellVolume #molecule^-1 sec^-1\n",
    "    k_off = 0.55 #sec^-1\n",
    "    \n",
    "    helicase_removal_rate = 600 #s^-1\n",
    "    \n",
    "    nonOricSpec = ['High_Affinity_Site','High_Affinity_Bound','High_Affinity_Site_oriC','High_Affinity_Bound_oriC','Low_Affinity_Site_1','Low_Affinity_Site_2','Low_Affinity_Bound_1','Low_Affinity_Bound_2']\n",
    "    sim.defineSpecies(nonOricSpec)\n",
    "    sim.addParticles(species = 'High_Affinity_Site', count = 16)\n",
    "    sim.addParticles(species = 'High_Affinity_Bound', count = 0)\n",
    "    sim.addReaction(('High_Affinity_Site', 'M_DnaA_c'),'High_Affinity_Bound',k_high)\n",
    "    \n",
    "    sim.addParticles(species = 'High_Affinity_Site_oriC', count = 1)\n",
    "    sim.addParticles(species = 'Low_Affinity_Site_1', count = 0)\n",
    "    sim.addParticles(species = 'Low_Affinity_Site_2', count = 0)\n",
    "    \n",
    "    sim.addReaction(('High_Affinity_Site_oriC', 'M_DnaA_c'),('High_Affinity_Bound_oriC','Low_Affinity_Site_1'),k_high)\n",
    "    sim.addReaction(('Low_Affinity_Site_1', 'M_DnaA_c'),('Low_Affinity_Bound_1','Low_Affinity_Site_2'),k_low)\n",
    "    sim.addReaction(('Low_Affinity_Site_2', 'M_DnaA_c'),('Low_Affinity_Bound_2','ssDNAunboundSite_1'),k_low)\n",
    "\n",
    "    species = []\n",
    "\n",
    "    for i in range(30):         #loop adds 30 terms for unbound sites\n",
    "        term = 'ssDNAunboundSite_'\n",
    "        term = term + str(i+1)\n",
    "        species.append(term)\n",
    "    for j in range(30):         #adds 30 more terms of Bound sites\n",
    "        bnd = 'ssDNABoundSite_'\n",
    "        bnd = bnd + str(j+1)\n",
    "        species.append(bnd)\n",
    "    for k in range(30):\n",
    "        unbnd = 'ssDNA_Unbinding_'\n",
    "        unbnd = unbnd + str(k+1)\n",
    "        species.append(unbnd)\n",
    "        \n",
    "    species.append('Initiator')\n",
    "\n",
    "    sim.defineSpecies(species)\n",
    "\n",
    "    for specy in species:\n",
    "        sim.addParticles(species = specy, count = 0) #gives all particles count of 0 \n",
    "\n",
    "    sim.addParticles(species = 'ssDNAunboundSite_1', count = 0) #starts at 1(only site existing @ start\n",
    "    \n",
    "    # Bidning and unbinding reactions are constructed so that DnaA in the middle of the filament cannot unbind and\n",
    "    # DnaA only bind next to the last DnaA in the filament.\n",
    "\n",
    "    # Add the binding reactions for each of the 30 binding events in filament formation.\n",
    "    for i in range (1,30):\n",
    "        sim.addReaction(('ssDNAunboundSite_' + str(i),'M_DnaA_c'), ('ssDNAunboundSite_' + str(i+1),'ssDNABoundSite_' + str(i)),k_on)\n",
    "        sim.addReaction(('ssDNAunboundSite_' + str(i+1), 'ssDNABoundSite_' + str(i)),('ssDNAunboundSite_' + str(i),'M_DnaA_c'),k_off)\n",
    "\n",
    "    sim.addReaction(('ssDNAunboundSite_30', 'M_DnaA_c'),('ssDNABoundSite_30','ssDNA_Unbinding_30','Initiator'),k_on)\n",
    "    \n",
    "    # Add the unbinding reactions for each of the 30 possible unbinding events in filament formation.\n",
    "    for i in range (2,31):\n",
    "        sim.addReaction(('ssDNA_Unbinding_' + str(i), 'ssDNABoundSite_' + str(i)),('ssDNA_Unbinding_' + str(i-1),'M_DnaA_c'),helicase_removal_rate)\n",
    "        \n",
    "    sim.addReaction(('ssDNA_Unbinding_1', 'ssDNABoundSite_1'),'M_DnaA_c',helicase_removal_rate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add replication reaction to duplicate each gene in the model.\n",
    "# The model is constucted so that replication cannot occur before completion of replication initiation.\n",
    "# The species \"chromosome\" is used to ensure the genome is only duplicated once.  It is a placeholder to say\n",
    "# the original genome is duplicated.\n",
    "\n",
    "def addReplication():\n",
    "    Total_genes = 0\n",
    "    K0rep = 0.26e-3\n",
    "    KDrep = 0.001\n",
    "    kcatrep = 600\n",
    "    datp = 0.009*2 # mM\n",
    "    dttp = 0.011*2 # mM\n",
    "    dctp = 0.006*2 # mM\n",
    "    dgtp = 0.0035*2 # mM\n",
    "    DNApol3 = 35*countToMiliMol # mM\n",
    "    RepRate = kcatrep/((1+K0rep/DNApol3)*(KDrep**2/datp/dttp) + 203606*KDrep/2/datp + 207816*KDrep/2/dttp + 67238*KDrep/2/dctp + 64720*KDrep/2/dgtp + 271690)\n",
    "    print(RepRate)\n",
    "    fastcalc = 600/270000\n",
    "    print(fastcalc)\n",
    "    Genes = ['Initiator']\n",
    "    for specy in ModelSpecies:\n",
    "        if 'gene' in specy:\n",
    "            Total_genes = Total_genes + 1\n",
    "            Genes.append(specy)\n",
    "    for i in range(271690):\n",
    "        # Since the length is half the genome, \n",
    "        Genes.append('ATP_rep')\n",
    "        Genes.append('ATP_rep')\n",
    "    print(Total_genes)\n",
    "    chromosome = ['chromosome']\n",
    "    sim.defineSpecies(chromosome)\n",
    "    sim.addParticles(species = 'chromosome', count = 1)\n",
    "    sim.addReaction(('chromosome', 'Initiator'), tuple(Genes), RepRate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Populate the model with particles and reactions.\n",
    "\n",
    "# Each ATP cost is specifically labelled in the model for the process.\n",
    "# ATP is written as a \"product\", but this is just a counter for the number of hydrolysis events in each process.\n",
    "ATP_costs = ['ATP_rep','ATP_trsc','ATP_translat','ATP_mRNAdeg','ATP_ptndeg']\n",
    "\n",
    "sim.defineSpecies(ATP_costs)\n",
    "\n",
    "genes_in_model = []\n",
    "\n",
    "for index, row in PtnMetDF.iterrows():\n",
    "    addNamedPtn(row[\"species\"], row[\"gene\"])\n",
    "    \n",
    "for index, row in riboPtnMetDF.iterrows():\n",
    "    addRiboPtn(row[\"species\"], row[\"gene\"])\n",
    "#     result = addPtns(row[\"species\"], row[\"gene\"])\n",
    "#     if result:\n",
    "#         addedPtns.append( result )\n",
    "\n",
    "# for gene in MetPtnGenes:\n",
    "#     addMetPtn(gene)\n",
    "\n",
    "# for index, row in RNApolPtnMetDF.iterrows():\n",
    "#     addPtn(row[\"species\"], row[\"gene\"])    \n",
    "    \n",
    "for gene in genomePtnLocDict:\n",
    "    if gene not in genes_in_model:\n",
    "        addPtn(gene)\n",
    "    \n",
    "for index, row in trnaMetDF.iterrows():\n",
    "    addtRNA(row[\"species\"], row[\"gene\"])\n",
    "    \n",
    "addrRNA()\n",
    "\n",
    "addRepInit()\n",
    "addReplication()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Print out all particles and reactions in the simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sim\n",
    "# Uncomment with \"ctrl /\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Stop evaluation here if you want to wait to run the simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import stop_evaluation \n",
    "#Uncomment the line above with \"ctrl /\" if you want the notebook to stop evaluation before running the simulation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run the Simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up the times\n",
    "simtime = 6300 # s = 105 min\n",
    "sim.setTimestep(ms(100))\n",
    "sim.setWriteInterval(1)\n",
    "sim.setSimulationTime(simtime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save simulation as LM file.\n",
    "\n",
    "filename='MinCellGeneProc_allGenes_ATP.lm'\n",
    "os.system(\"rm -rf %s\"%(filename))\n",
    "sim.save(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Choose number of replicate cells to simulate.\n",
    "reps = 10\n",
    "sim.run(filename, \"lm::cme::GillespieDSolver\", reps)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot Simulation Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Retrieve times at which DnaA is first bound at site 30 \n",
    "#store data in a list to then create histogram\n",
    "\n",
    "BindTimes = []\n",
    "\n",
    "fh = openLMFile(filename)\n",
    "for i in range(reps):\n",
    "    trace = getSpecieTrace(fh, 'Initiator', replicate = i+1)\n",
    "#     'ssDNABoundSite_30'\n",
    "    if trace.any() == 0:\n",
    "        print('Replicate',i+1,'did not bind any DnaA at site 30___!')\n",
    "        print()\n",
    "    else:\n",
    "        BindTimes.append(np.where(trace==1)[0][0])\n",
    "closeLMFile(fh)\n",
    "\n",
    "BindTimesMin = []\n",
    "for time in BindTimes:\n",
    "    sec_to_min = time / 60\n",
    "    BindTimesMin.append(sec_to_min)\n",
    "\n",
    "# print(BindTimes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "plt.figure(figsize=(8,6))\n",
    "sns.distplot(BindTimesMin)\n",
    "# plt.title('Probability Distribution of 30th DnaA Binding Times')\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Probability Density to Bind 30th DnaA')\n",
    "\n",
    "totalTimes=0\n",
    "for bindTime in BindTimesMin:\n",
    "    totalTimes = totalTimes + bindTime\n",
    "print(totalTimes/len(BindTimesMin))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "repnums = [1,2,3,4] # set the replicate number to be graphed here\n",
    "\n",
    "def SumBoundsites(rep): # input for function is the replicate number\n",
    "    base = []  # creates empty list\n",
    "\n",
    "    for x in range(simtime+1): # fills list with zeros, with length equal to total time \n",
    "        base.append(0) \n",
    "    \n",
    "    for i in range(30):  #loop 30 times (for each binding site)\n",
    "        specName = 'ssDNABoundSite_' + str(i+1)  \n",
    "        trace = getSpecieTrace(fh, specName, replicate = rep)  # gets the data for binding site i\n",
    "        base = base + trace # comes in the form of a list so just sum it with the components of the base list\n",
    "    return base  \n",
    "\n",
    "\n",
    "fh = openLMFile(filename)\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "for rep in repnums:\n",
    "    totalDnaA = SumBoundsites(rep)\n",
    "    time = np.arange(0,simtime+1,1) \n",
    "    plt.plot(time/60,totalDnaA)\n",
    "    \n",
    "# title = 'Total Number of DnaA bound for Replicate ' + str(repnum)\n",
    "# plt.title(title)\n",
    "plt.xlim(0,40)\n",
    "plt.ylabel('DnaA Filament Length (number of DnaA)')\n",
    "plt.xlabel('Time (min)')\n",
    "\n",
    "closeLMFile(fh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Definitions to Plot Single Replicates and Averages"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Randomly sample 10 single replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PlotSpecies(species):\n",
    "    fileHandle=openLMFile(filename)\n",
    "    for i in range(1,min(reps,10)):\n",
    "#         plotTraceFromFile(filename, [species], i, species + \".%d.png\"%(i))\n",
    "        plotTrace(fileHandle, [species],i)\n",
    "        plt.show()\n",
    "    closeLMFile(fileHandle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PlotSpecies(\"M_RNA_JCVISYN3A_0445_c\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot of a single replicate of gene 0445; glucose-6-phosphate isomerase.\n",
    "# Note due to stochasticity, this plot will be completely different for every replicate and different among rerunning the sim.\n",
    "\n",
    "fileHandle=openLMFile(filename)\n",
    "specName = \"M_RNA_JCVISYN3A_0445_c\"\n",
    "# Choose a replicate number to plot.\n",
    "replicat = 10\n",
    "trace = getSpecieTrace(fileHandle, specName, replicate = replicat)\n",
    "time = np.arange(0,simtime+1,1)\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "plt.figure(figsize=(8,6))\n",
    "plt.plot(time/60,trace)\n",
    "plt.ylabel('Single Replicate mRNA Abundance')\n",
    "plt.xlabel('Time (min)')\n",
    "closeLMFile(fileHandle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a definition to average each replicate from the simulation at each time point.\n",
    "\n",
    "def AverageSpec(specName):\n",
    "    fileHandle=openLMFile(filename)\n",
    "    simTotal = np.zeros(simtime+1)\n",
    "    simAvg = []\n",
    "    \n",
    "    for rep in range(1,reps+1):\n",
    "#         base = []\n",
    "#         for x in range(6000): # fills list with zeros, with length equal to total time \n",
    "#             base.append(0)\n",
    "        trace = getSpecieTrace(fileHandle, specName, replicate = rep)\n",
    "#         base = base + trace # comes in the form of a list so just sum it with the components of the base list\n",
    "        simTotal += trace\n",
    "        \n",
    "    for index in simTotal:\n",
    "        index_avg = index/reps\n",
    "        simAvg.append(index_avg)\n",
    "        \n",
    "#     closeLMFile(filename)\n",
    "    \n",
    "    return simAvg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Stop evaluation here if you don't want to immediately make every plot.\n",
    "### Warning: For 1000 replicates, averaging the data can take hours if you make every plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import stop_evaluation \n",
    "#Uncomment or comment out the line above with \"ctrl /\"\" if you want the notebook to stop evaluation before making plots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot of average gene duplication (used gene 0067).  There are a few other species to look at if interested.\n",
    "\n",
    "time = np.arange(0,simtime+1,1) # list of time steps, for x axis\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "plt.figure(figsize=(8,6))\n",
    "# plt.plot(time,AverageSpec(\"ssDNABoundSite_30\"),'g')\n",
    "plt.plot(time/60,AverageSpec('JCVISYN3A_0067_gene'),'b')\n",
    "# plt.plot(time/60,AverageSpec('M_RNA_JCVISYN3A_0001_c'), 'g')\n",
    "# plt.plot(time/60,AverageSpec('M_DnaA_c'), 'g')\n",
    "# plt.plot(time/60,AverageSpec('ssDNAunboundSite_29'), 'g')\n",
    "# plt.title('Average')\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average DNA Copy Number')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot of average free (unbound) DnaA count.\n",
    "\n",
    "time = np.arange(0,simtime+1,1) # list of time steps, for x axis\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "plt.plot(time/60,AverageSpec('M_DnaA_c'), 'darkorange')\n",
    "\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average Abundance of Free DnaA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rrna_spec = ['M_rRNA_5S_c','M_rRNA_16S_c','M_rRNA_23S_c']\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "time = np.arange(0,simtime+1,1)\n",
    "\n",
    "for rRNA in rrna_spec:\n",
    "    plt.plot(time/60,AverageSpec(rRNA),'r')\n",
    "    \n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('rRNA Generated')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Average mRNA counts for ribosomal protein coding mRNA.\n",
    "\n",
    "metListmRNA = []\n",
    "\n",
    "# for rnaID in ModelSpecies:\n",
    "#     if \"RNA_MMS\" in rnaID:\n",
    "#         metListmRNA.append(rnaID)\n",
    "\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "time = np.arange(0,simtime+1,1)\n",
    "\n",
    "# for mmcode in gene_list_other:\n",
    "#     mRNA = \"M_RNA_\" + mmcode + \"_c\"\n",
    "#     metListmRNA.append(mRNA)\n",
    "#     plt.plot(time/60,AverageSpec(mRNA),'y')\n",
    "\n",
    "for index, row in riboPtnMetDF.iterrows():\n",
    "    mmcode = row[\"gene\"]\n",
    "    LocusNum = mmcode.split('_')[1]\n",
    "    JCVI3AID = \"JCVISYN3A_\" + LocusNum\n",
    "    mRNA = \"M_RNA_\" + JCVI3AID + \"_c\"\n",
    "    metListmRNA.append(mRNA)\n",
    "    avgTrace = AverageSpec(mRNA)\n",
    "#     plt.plot(time/60,AverageSpec(mRNA),'crimson')\n",
    "    plt.plot(time/60,avgTrace,'crimson')\n",
    "#     if avgTrace[-1] > 4:\n",
    "#         print(mRNA,':',avgTrace[-1])\n",
    "#     if avgTrace[-1] < 1.5:\n",
    "#         print(mRNA,':',avgTrace[-1])\n",
    "        \n",
    "    \n",
    "# metListmRNA.append(\"M_RNA_JCVISYN3A_0001_c\")\n",
    "    \n",
    "\n",
    "        \n",
    "# for index, row in riboPtnMetDF.iterrows():\n",
    "#     mmcode = row[\"gene\"]\n",
    "#     LocusNum = mmcode.split('_')[1]\n",
    "#     JCVI3AID = \"JCVISYN3A_\" + LocusNum\n",
    "#     mRNA = \"M_RNA_\" + JCVI3AID + \"_c\"\n",
    "# #     metListmRNA.append(mRNA)\n",
    "#     plt.plot(time/60,AverageSpec(mRNA),'crimson')\n",
    "        \n",
    "# for mmcode in MetPtnGenes:\n",
    "#     LocusNum = mmcode.split('_')[1]\n",
    "#     JCVI3AID = \"JCVISYN3A_\" + LocusNum\n",
    "#     mRNA = \"M_RNA_\" + JCVI3AID + \"_c\"\n",
    "# #     metListmRNA.append(mRNA)\n",
    "#     plt.plot(time/60,AverageSpec(mRNA),'g')\n",
    "        \n",
    "# plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0001_c\"),'darkorange',linewidth = 4)\n",
    "    \n",
    "# print(len(metListmRNA))\n",
    "\n",
    "# time = np.arange(0,simtime+1,1) # list of time steps, for x axis\n",
    "\n",
    "# for mrna in metListmRNA:\n",
    "#     plt.plot(time/60,AverageSpec(mrna))\n",
    "\n",
    "# plt.title('Average mRNA Abundance')\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average mRNA Abundance')\n",
    "    \n",
    "# metListPtns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Average mRNA counts for metabolic protein coding mRNA.\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "time = np.arange(0,simtime+1,1)\n",
    "\n",
    "metListmRNA.append(\"M_RNA_JCVISYN3A_0065_c\")\n",
    "plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0065_c\"),'g')\n",
    "\n",
    "metListmRNA.append(\"M_RNA_JCVISYN3A_0621_c\")\n",
    "plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0621_c\"),'g')\n",
    "\n",
    "\n",
    "for mmcode in MetPtnGenes:\n",
    "    LocusNum = mmcode.split('_')[1]\n",
    "    JCVI3AID = \"JCVISYN3A_\" + LocusNum\n",
    "    mRNA = \"M_RNA_\" + JCVI3AID + \"_c\"\n",
    "    metListmRNA.append(mRNA)\n",
    "    plt.plot(time/60,AverageSpec(mRNA),'g')\n",
    "#     plt.plot(time/60,AverageSpec(mRNA),'g')\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average mRNA Abundance')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Average mRNA counts for unclear function protein coding mRNA.\n",
    "\n",
    "unclear = 0\n",
    "\n",
    "NEWmRNA = []\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "time = np.arange(0,simtime+1,1)\n",
    "\n",
    "NEWmRNA.append('M_RNA_JCVISYN3A_0918_c')\n",
    "NEWmRNA.append('M_RNA_JCVISYN3A_0913_c')\n",
    "NEWmRNA.append('M_RNA_JCVISYN3A_0930_c')\n",
    "NEWmRNA.append('M_RNA_JCVISYN3A_0931_c')\n",
    "NEWmRNA.append('M_RNA_JCVISYN3A_0932_c')\n",
    "NEWmRNA.append('M_RNA_JCVISYN3A_0154_c')\n",
    "\n",
    "for rnaID in NEWmRNA:\n",
    "    metListmRNA.append(rnaID)\n",
    "    plt.plot(time/60,AverageSpec(rnaID),'gray')\n",
    "\n",
    "for rnaID in ModelSpecies:\n",
    "    if (\"RNA_JCVI\" in rnaID) and (rnaID not in metListmRNA):\n",
    "        \n",
    "        locus_code = rnaID.split('3A_')[1].replace('_c','')\n",
    "        mmcode = 'MMSYN1_' + locus_code\n",
    "\n",
    "        function = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 10].strip()\n",
    "        if 'clear' in function:\n",
    "            unclear = unclear + 1\n",
    "\n",
    "            metListmRNA.append(rnaID)\n",
    "            plt.plot(time/60,AverageSpec(rnaID),'gray')\n",
    "\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average mRNA Abundance')\n",
    "            \n",
    "print(unclear)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Average mRNA counts for genetic information processing and cell division protein coding mRNA.\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "time = np.arange(0,simtime+1,1)\n",
    "\n",
    "metListmRNA.append('M_RNA_JCVISYN3A_0001_c')\n",
    "\n",
    "for rnaID in ModelSpecies:\n",
    "    if (\"RNA_JCVI\" in rnaID) and (rnaID not in metListmRNA):\n",
    "        metListmRNA.append(rnaID)\n",
    "        plt.plot(time/60,AverageSpec(rnaID),'dodgerblue')\n",
    "        \n",
    "plt.plot(time/60,AverageSpec('M_RNA_JCVISYN3A_0001_c'),'darkorange',linewidth = 4)\n",
    "        \n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average mRNA Abundance')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metListmRNA = []\n",
    "\n",
    "# for rnaID in ModelSpecies:\n",
    "#     if \"RNA_MMS\" in rnaID:\n",
    "#         metListmRNA.append(rnaID)\n",
    "\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "time = np.arange(0,simtime+1,1)\n",
    "\n",
    "# for mmcode in gene_list_other:\n",
    "#     mRNA = \"M_RNA_\" + mmcode + \"_c\"\n",
    "#     metListmRNA.append(mRNA)\n",
    "#     plt.plot(time/60,AverageSpec(mRNA),'y')\n",
    "\n",
    "for index, row in riboPtnMetDF.iterrows():\n",
    "    mmcode = row[\"gene\"]\n",
    "    LocusNum = mmcode.split('_')[1]\n",
    "    JCVI3AID = \"JCVISYN3A_\" + LocusNum\n",
    "    mRNA = \"M_RNA_\" + JCVI3AID + \"_c\"\n",
    "    metListmRNA.append(mRNA)\n",
    "#     plt.plot(time/60,AverageSpec(mRNA),'crimson')\n",
    "        \n",
    "for mmcode in MetPtnGenes:\n",
    "    LocusNum = mmcode.split('_')[1]\n",
    "    JCVI3AID = \"JCVISYN3A_\" + LocusNum\n",
    "    mRNA = \"M_RNA_\" + JCVI3AID + \"_c\"\n",
    "    metListmRNA.append(mRNA)\n",
    "#     plt.plot(time/60,AverageSpec(mRNA),'g')\n",
    "\n",
    "\n",
    "    \n",
    "metListmRNA.append(\"M_RNA_JCVISYN3A_0001_c\")\n",
    "\n",
    "    \n",
    "for rnaID in ModelSpecies:\n",
    "    if (\"RNA_JCVI\" in rnaID) and (rnaID not in metListmRNA):\n",
    "        metListmRNA.append(rnaID)\n",
    "#         plt.plot(time/60,AverageSpec(rnaID),'dodgerblue')\n",
    "        \n",
    "# plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0001_c\"),'darkorange',linewidth = 4)\n",
    "    \n",
    "print(len(metListmRNA))\n",
    "\n",
    "Total_mRNA = np.zeros(6301)\n",
    "\n",
    "for mRNA in metListmRNA:\n",
    "    avg_abund = AverageSpec(mRNA)\n",
    "#     print(mRNA,' : ',avg_abund[-1])\n",
    "    Total_mRNA = Total_mRNA + avg_abund\n",
    "    \n",
    "plt.plot(time/60,Total_mRNA)\n",
    "\n",
    "# time = np.arange(0,simtime+1,1) # list of time steps, for x axis\n",
    "\n",
    "# for mrna in metListmRNA:\n",
    "#     plt.plot(time/60,AverageSpec(mrna))\n",
    "\n",
    "# plt.title('Average mRNA Abundance')\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average Total mRNA Abundance')\n",
    "    \n",
    "# metListPtns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "time = np.arange(0,simtime+1,1) # list of time steps, for x axis\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "metListtRNAmulti = []\n",
    "\n",
    "trna_met_avg = AverageSpec('M_trnamet_c')\n",
    "plt.plot(time/60,trna_met_avg - trna_met_avg[0],'k')\n",
    "metListtRNAmulti.append('M_trnamet_c')\n",
    "\n",
    "trna_leu_avg = AverageSpec('M_trnaleu_c')\n",
    "plt.plot(time/60,trna_leu_avg - trna_leu_avg[0],'k')\n",
    "metListtRNAmulti.append('M_trnaleu_c')\n",
    "\n",
    "trna_thr_avg = AverageSpec('M_trnathr_c')\n",
    "plt.plot(time/60,trna_thr_avg - trna_thr_avg[0],'c')\n",
    "metListtRNAmulti.append('M_trnathr_c')\n",
    "\n",
    "trna_trp_avg = AverageSpec('M_trnatrp_c')\n",
    "plt.plot(time/60,trna_trp_avg - trna_trp_avg[0],'c')\n",
    "metListtRNAmulti.append('M_trnatrp_c')\n",
    "\n",
    "trna_lys_avg = AverageSpec('M_trnalys_c')\n",
    "plt.plot(time/60,trna_lys_avg - trna_lys_avg[0],'c')\n",
    "metListtRNAmulti.append('M_trnalys_c')\n",
    "\n",
    "trna_arg_avg = AverageSpec('M_trnaarg_c')\n",
    "plt.plot(time/60,trna_arg_avg - trna_arg_avg[0],'c')\n",
    "metListtRNAmulti.append('M_trnaarg_c')\n",
    "\n",
    "trna_ser_avg = AverageSpec('M_trnaser_c')\n",
    "plt.plot(time/60,trna_ser_avg - trna_ser_avg[0],'c')\n",
    "metListtRNAmulti.append('M_trnaser_c')\n",
    "\n",
    "# metListtRNAmulti\n",
    "\n",
    "metListtRNAsingle = []\n",
    "\n",
    "for rnaID in ModelSpecies:\n",
    "    if (\"M_trna\" in rnaID) and (rnaID not in metListtRNAmulti):\n",
    "        metListtRNAsingle.append(rnaID)\n",
    "        \n",
    "for trna in metListtRNAsingle:\n",
    "    trna_avg = AverageSpec(trna)\n",
    "    plt.plot(time/60,trna_avg - trna_avg[0],'m')\n",
    "    \n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average tRNA Generated')\n",
    "        \n",
    "# metListtRNAsingle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Total_trna = np.zeros(6301)\n",
    "\n",
    "time = np.arange(0,simtime+1,1) # list of time steps, for x axis\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "for rnaID in ModelSpecies:\n",
    "    \n",
    "    if \"M_trna\" in rnaID:\n",
    "        trna_avg = AverageSpec(trna)\n",
    "        trna_formed = trna_avg - trna_avg[0]\n",
    "        Total_trna = Total_trna + trna_formed\n",
    "        \n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average Total tRNA Generated')\n",
    "           \n",
    "plt.plot(time/60,Total_trna,'k')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot Protein Abundances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metListPtns = []\n",
    "\n",
    "UnkPtns = []\n",
    "\n",
    "metPtns = []\n",
    "\n",
    "OtherPtns = []\n",
    "\n",
    "NEWptn = []\n",
    "\n",
    "# for ptnID in ModelSpecies:\n",
    "#     if \"PTN_JCVI\" in ptnID:\n",
    "#         metListPtns.append(ptnID)\n",
    "        \n",
    "time = np.arange(0,simtime+1,1) # list of time steps, for x axis\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "Low_proteomics = 0\n",
    "\n",
    "for mmcode in MetPtnGenes:\n",
    "    LocusNum = mmcode.split('_')[1]\n",
    "    JCVI3AID = \"JCVISYN3A_\" + LocusNum\n",
    "    ptn = \"M_PTN_\" + JCVI3AID + \"_c\"\n",
    "    \n",
    "    metPtns.append(ptn)\n",
    "        \n",
    "metListPtns.append('M_trdox_c')\n",
    "metPtns.append('M_trdox_c')\n",
    "\n",
    "metListPtns.append('M_apoACP_c')\n",
    "metPtns.append('M_apoACP_c')\n",
    "\n",
    "NEWptn.append('M_PTN_JCVISYN3A_0918_c')\n",
    "NEWptn.append('M_PTN_JCVISYN3A_0913_c')\n",
    "NEWptn.append('M_PTN_JCVISYN3A_0930_c')\n",
    "NEWptn.append('M_PTN_JCVISYN3A_0931_c')\n",
    "NEWptn.append('M_PTN_JCVISYN3A_0932_c')\n",
    "NEWptn.append('M_PTN_JCVISYN3A_0154_c')\n",
    "\n",
    "for ptnID in NEWptn:\n",
    "    metListPtns.append(ptnID)\n",
    "    \n",
    "    UnkPtns.append(ptnID)\n",
    "\n",
    "for ptnID in ModelSpecies:\n",
    "    if (\"PTN_JCVI\" in ptnID) and (ptnID not in metListPtns):\n",
    "        \n",
    "        locus_code = ptnID.split('3A_')[1].replace('_c','')\n",
    "        mmcode = 'MMSYN1_' + locus_code\n",
    "\n",
    "        function = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 10].strip()\n",
    "        \n",
    "        if 'clear' in function:\n",
    "            \n",
    "            UnkPtns.append(ptnID)\n",
    "            \n",
    "            metListPtns.append(ptnID)\n",
    "            \n",
    "                \n",
    "for ptnID in ModelSpecies:\n",
    "    if (\"PTN_JCVI\" in ptnID) and (ptnID not in metListPtns):\n",
    "        metListPtns.append(ptnID)\n",
    "            \n",
    "        avgTrace = AverageSpec(ptnID)\n",
    "    \n",
    "        proteomics = avgTrace[0]\n",
    "\n",
    "        if proteomics > 10:\n",
    "            scaled_abundance = avgTrace/proteomics\n",
    "            plt.plot(time/60,scaled_abundance,'b')\n",
    "\n",
    "        else:\n",
    "            Low_proteomics = Low_proteomics + 1\n",
    "            \n",
    "for ptn in UnkPtns:\n",
    "    avgTrace = AverageSpec(ptn)\n",
    "    proteomics = avgTrace[0]\n",
    "    \n",
    "    if proteomics > 10:\n",
    "        scaled_abundance = avgTrace/proteomics\n",
    "        plt.plot(time/60,scaled_abundance,'gray')\n",
    "        \n",
    "    else:\n",
    "        Low_proteomics = Low_proteomics + 1\n",
    "        \n",
    "for ptn in metPtns:\n",
    "    avgTrace = AverageSpec(ptn)\n",
    "    proteomics = avgTrace[0]\n",
    "    \n",
    "    if proteomics > 10:\n",
    "        scaled_abundance = avgTrace/proteomics\n",
    "        plt.plot(time/60,scaled_abundance,'g')\n",
    "        \n",
    "    else:\n",
    "        Low_proteomics = Low_proteomics + 1\n",
    "            \n",
    "metListPtns.append('M_DnaA_c')\n",
    "avgTrace = AverageSpec('M_DnaA_c')\n",
    "proteomics = avgTrace[0]\n",
    "scaled_abundance = avgTrace/proteomics\n",
    "plt.plot(time/60,scaled_abundance,'b')\n",
    "        \n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average Scaled Protein Abundance')\n",
    "\n",
    "print(len(metListPtns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PtnDistList = []\n",
    "PtnDist = []\n",
    "\n",
    "for ptnID in ModelSpecies:\n",
    "    if \"PTN_JCVI\" in ptnID:\n",
    "        PtnDistList.append(ptnID)\n",
    "\n",
    "for index, row in PtnMetDF.iterrows():\n",
    "    result = row[\"species\"]\n",
    "    if result:\n",
    "        PtnDistList.append( result )\n",
    "        \n",
    "for ptn in PtnDistList:\n",
    "    avgTrace = AverageSpec(ptn)\n",
    "    \n",
    "    proteomics = avgTrace[0]\n",
    "    \n",
    "    if proteomics > 10:\n",
    "        scaledAvg = avgTrace[-1]/avgTrace[0]\n",
    "        PtnDist.append(scaledAvg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Histogram of the protein counts scaled to their proteomics count at 105 minutes.\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "plt.figure(figsize=(8,6))\n",
    "plt.hist(PtnDist, bins=30)\n",
    "# plt.title('Probability Distribution of 30th DnaA Binding Times')\n",
    "plt.xlabel('Scaled Protein Abundance After 105 min')\n",
    "plt.ylabel('Number of Unique Proteins')\n",
    "plt.xlim(0,10.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print('Proteins with proteomics below 10 or no proteomics: ', Low_proteomics)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Print out details for proteins underproduced, overproduced, and just right."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ANALYSIS OF UNDERPRODUCED PROTEINS\n",
    "\n",
    "metListPtns = []\n",
    "\n",
    "for ptnID in ModelSpecies:\n",
    "    if \"PTN_JCVI\" in ptnID:\n",
    "        metListPtns.append(ptnID)\n",
    "\n",
    "for index, row in PtnMetDF.iterrows():\n",
    "    result = row[\"species\"]\n",
    "    if result:\n",
    "        metListPtns.append( result )\n",
    "# metListPtns.append('M_trdox_c')        \n",
    "# metListPtns.remove('M_PTN_MMSYN1_0689_c')\n",
    "# print(metIDListRiboPtn)\n",
    "        \n",
    "time = np.arange(0,simtime+1,1) # list of time steps, for x axis\n",
    "\n",
    "# plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "# plt.figure(figsize=(8,6))\n",
    "\n",
    "lowPtns = 0\n",
    "\n",
    "for ptn in metListPtns:\n",
    "    avgTrace = AverageSpec(ptn)\n",
    "    \n",
    "    proteomics = avgTrace[0]\n",
    "    \n",
    "    if proteomics > 10:\n",
    "        scaled_abundance = avgTrace/proteomics\n",
    "#         plt.plot(time/60,scaled_abundance)\n",
    "#         if scaled_abundance[-1] > 4:\n",
    "#             print(ptn,proteomics,scaled_abundance[-1])\n",
    "            \n",
    "        if scaled_abundance[-1]<1.5:\n",
    "            lowPtns = lowPtns + 1\n",
    "            locus_code = ptn.split('3A_')[1].replace('_c','')\n",
    "            mmcode = 'MMSYN1_' + locus_code\n",
    "            # Checks if a translation to JCVISYN2* code is available\n",
    "\n",
    "            ptnName = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 8].strip()\n",
    "            GeneLength = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 2]\n",
    "            function = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 10].strip()\n",
    "            \n",
    "                \n",
    "            print(lowPtns,': ',ptnName)\n",
    "            print('Function: ',function)\n",
    "            print('Gene Length: ',GeneLength)\n",
    "            print('Model ID: ',ptn)\n",
    "            print('Proteomics: ',proteomics)\n",
    "            print('Scaled Abundance after 105 min: ',scaled_abundance[-1])\n",
    "            print(ptn,proteomics,scaled_abundance[-1])\n",
    "            print()\n",
    "\n",
    "# plt.title('Average ptn Pop')\n",
    "# plt.xlabel('Time (min)')\n",
    "# plt.ylabel('Average Scaled Protein Abundance')\n",
    "    \n",
    "# metListPtns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# ANALYSIS OF OVERPRODUCED PROTEINS\n",
    "\n",
    "metListPtns = []\n",
    "\n",
    "for ptnID in ModelSpecies:\n",
    "    if \"PTN_JCVI\" in ptnID:\n",
    "        metListPtns.append(ptnID)\n",
    "\n",
    "for index, row in PtnMetDF.iterrows():\n",
    "    result = row[\"species\"]\n",
    "    if result:\n",
    "        metListPtns.append( result )\n",
    "# metListPtns.append('M_trdox_c')        \n",
    "# metListPtns.remove('M_PTN_MMSYN1_0689_c')\n",
    "# print(metIDListRiboPtn)\n",
    "        \n",
    "time = np.arange(0,simtime+1,1) # list of time steps, for x axis\n",
    "\n",
    "# plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "# plt.figure(figsize=(8,6))\n",
    "\n",
    "highPtns = 0\n",
    "\n",
    "for ptn in metListPtns:\n",
    "    avgTrace = AverageSpec(ptn)\n",
    "    \n",
    "    proteomics = avgTrace[0]\n",
    "    \n",
    "    if proteomics > 10:\n",
    "        scaled_abundance = avgTrace/proteomics\n",
    "#         plt.plot(time/60,scaled_abundance)\n",
    "        \n",
    "        if scaled_abundance[-1] > 4:\n",
    "            highPtns = highPtns + 1\n",
    "            \n",
    "            if 'JCVISYN3A' in ptn:\n",
    "                locus_code = ptn.split('3A_')[1].replace('_c','')\n",
    "                mmcode = 'MMSYN1_' + locus_code\n",
    "                # Checks if a translation to JCVISYN2* code is available\n",
    "\n",
    "                ptnName = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 8].strip()\n",
    "                GeneLength = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 2]\n",
    "                function = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 10].strip()\n",
    "            else:\n",
    "                GeneLength = ''\n",
    "                ptnName = ''\n",
    "                \n",
    "            print(highPtns,': ',ptnName)\n",
    "            print('Function: ',function)\n",
    "            print('Model ID: ',ptn)\n",
    "            print('Gene Length: ',GeneLength)\n",
    "            print('Proteomics: ',proteomics)\n",
    "            print('Scaled Abundance after 105 min: ',scaled_abundance[-1])\n",
    "            print()\n",
    "            \n",
    "#         if scaled_abundance[-1]<1.5:\n",
    "#             print(ptn,proteomics,scaled_abundance[-1])\n",
    "#     else:\n",
    "#         print(ptn,avgTrace[0],avgTrace[-1]/avgTrace[0])\n",
    "\n",
    "# plt.title('Average ptn Pop')\n",
    "# plt.xlabel('Time (min)')\n",
    "# plt.ylabel('Average Scaled Protein Abundance')\n",
    "    \n",
    "    #trdox length = 309 nt\n",
    "    #apoACP length  = 222 nt\n",
    "    \n",
    "# metListPtns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ANALYSIS OF PROTEINS WITHIN 10% OF BEING DOUBLED\n",
    "\n",
    "metListPtns = []\n",
    "\n",
    "for ptnID in ModelSpecies:\n",
    "    if \"PTN_JCVI\" in ptnID:\n",
    "        metListPtns.append(ptnID)\n",
    "\n",
    "for index, row in PtnMetDF.iterrows():\n",
    "    result = row[\"species\"]\n",
    "    if result:\n",
    "        metListPtns.append( result )\n",
    "# metListPtns.append('M_trdox_c')        \n",
    "# metListPtns.remove('M_PTN_MMSYN1_0689_c')\n",
    "# print(metIDListRiboPtn)\n",
    "        \n",
    "time = np.arange(0,simtime+1,1) # list of time steps, for x axis\n",
    "\n",
    "# plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "# plt.figure(figsize=(8,6))\n",
    "\n",
    "highPtns = 0\n",
    "\n",
    "for ptn in metListPtns:\n",
    "    avgTrace = AverageSpec(ptn)\n",
    "    \n",
    "    proteomics = avgTrace[0]\n",
    "    \n",
    "    if proteomics > 10:\n",
    "        scaled_abundance = avgTrace/proteomics\n",
    "#         plt.plot(time/60,scaled_abundance)\n",
    "        \n",
    "        if (scaled_abundance[-1] > 1.7) and (scaled_abundance[-1] < 2.3):\n",
    "            highPtns = highPtns + 1\n",
    "            \n",
    "            if 'JCVISYN3A' in ptn:\n",
    "                locus_code = ptn.split('3A_')[1].replace('_c','')\n",
    "                mmcode = 'MMSYN1_' + locus_code\n",
    "                # Checks if a translation to JCVISYN2* code is available\n",
    "\n",
    "                ptnName = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 8].strip()\n",
    "                GeneLength = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 2]\n",
    "                function = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 10].strip()\n",
    "            else:\n",
    "                GeneLength = ''\n",
    "                ptnName = ''\n",
    "                \n",
    "            print(highPtns,': ',ptnName)\n",
    "            print('Function: ',function)\n",
    "            print('Model ID: ',ptn)\n",
    "            print('Gene Length: ',GeneLength)\n",
    "            print('Proteomics: ',proteomics)\n",
    "            print('Scaled Abundance after 105 min: ',scaled_abundance[-1])\n",
    "            print()\n",
    "            \n",
    "#         if scaled_abundance[-1]<1.5:\n",
    "#             print(ptn,proteomics,scaled_abundance[-1])\n",
    "#     else:\n",
    "#         print(ptn,avgTrace[0],avgTrace[-1]/avgTrace[0])\n",
    "\n",
    "# plt.title('Average ptn Pop')\n",
    "# plt.xlabel('Time (min)')\n",
    "# plt.ylabel('Average Scaled Protein Abundance')\n",
    "    \n",
    "    #trdox length = 309 nt\n",
    "    #apoACP length  = 222 nt\n",
    "    \n",
    "# metListPtns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Print out messenger counts for a few proteins that are interesting for various reasons.\n",
    "\n",
    "# Locus_of_int = ['0065','0621','0910','0238','0653','0662','0672','0607','0475','0290','0291','0804','0002','0216']\n",
    "\n",
    "# for tag in Locus_of_int:\n",
    "#     locusTag = 'JCVISYN3A_'+tag\n",
    "#     mrnaID = 'M_RNA_'+locusTag+'_c'\n",
    "#     avgTrace = AverageSpec(mrnaID)\n",
    "#     print(locusTag,' : ',avgTrace[-1])\n",
    "#     ptnID = 'M_PTN_'+locusTag+'_c'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get details for proteins in Central Metabolism.\n",
    "\n",
    "# CentMetList = ['0213','0220','0221','0227','0228','0229','0230','0233','0234','0262','0316','0435','0445','0451','0475','0493','0494','0495','0606','0607','0694','0726','0727','0729','0733','0779','0800','0831']\n",
    "\n",
    "# print(len(CentMetList))\n",
    "\n",
    "# CentMetListPtns = []\n",
    "\n",
    "# for locusNum in CentMetList:\n",
    "#     ptnID = 'M_PTN_JCVISYN3A_' + locusNum + '_c'\n",
    "#     CentMetListPtns.append(ptnID)\n",
    "\n",
    "# for ptn in CentMetListPtns:\n",
    "#     avgTrace = AverageSpec(ptn)\n",
    "    \n",
    "#     proteomics = avgTrace[0]\n",
    "    \n",
    "# #     if proteomics > 10:\n",
    "#     scaled_abundance = avgTrace/proteomics\n",
    "# #         plt.plot(time/60,scaled_abundance)\n",
    "            \n",
    "#     if 'JCVISYN3A' in ptn:\n",
    "#         locus_code = ptn.split('3A_')[1].replace('_c','')\n",
    "#         mmcode = 'MMSYN1_' + locus_code\n",
    "#                 # Checks if a translation to JCVISYN2* code is available\n",
    "\n",
    "#         ptnName = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 8].strip()\n",
    "#         GeneLength = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 2]\n",
    "#         function = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 10].strip()\n",
    "#     else:\n",
    "#         GeneLength = ''\n",
    "#         ptnName = ''\n",
    "        \n",
    "#     print(ptnName)            \n",
    "#     print('Function: ',function)\n",
    "#     print('Model ID: ',ptn)\n",
    "#     print('Gene Length: ',GeneLength)\n",
    "#     print('Proteomics: ',proteomics)\n",
    "#     print('Scaled Abundance after 105 min: ',scaled_abundance[-1])\n",
    "#     print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot production of ribosomal proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metIDListRiboPtn = []\n",
    "for index, row in riboPtnMetDF.iterrows():\n",
    "    result = row[\"species\"]\n",
    "    if result:\n",
    "        metIDListRiboPtn.append( result )\n",
    "# print(metIDListRiboPtn)\n",
    "\n",
    "time = np.arange(0,simtime+1,1) # list of time steps, for x axis\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "for ptn in metIDListRiboPtn:\n",
    "    avgTrace = AverageSpec(ptn)\n",
    "#     plt.plot(time/60,avgTrace-avgTrace[0])\n",
    "    plt.plot(time/60,avgTrace/avgTrace[0])\n",
    "#     print(ptn,': ',avgTrace[0],': ',avgTrace[-1]-avgTrace[0])\n",
    "#     print()\n",
    "\n",
    "# plt.title('Average ptn production')\n",
    "# plt.yscale('log')\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average Ribosomal Proteins Produced')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metIDListRiboPtn = []\n",
    "for index, row in riboPtnMetDF.iterrows():\n",
    "    result = row[\"species\"]\n",
    "    if result:\n",
    "        metIDListRiboPtn.append( result )\n",
    "print(metIDListRiboPtn)\n",
    "\n",
    "riboPtnDist = []\n",
    "RiboPtnProteomics = ['Proteomics Count']\n",
    "RiboPtnProd = ['Proteins Produced']\n",
    "\n",
    "for ptn in metIDListRiboPtn:\n",
    "    avgTrace = AverageSpec(ptn)\n",
    "    ptnProduced = avgTrace[-1]-avgTrace[0]\n",
    "    riboPtnDist.append(ptnProduced)\n",
    "    RiboPtnProd.append(ptnProduced)\n",
    "    RiboPtnProteomics.append(avgTrace[0])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Histogram of the number of each unique ribosomal protein produced.\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "plt.figure(figsize=(8,6))\n",
    "plt.hist(riboPtnDist, bins=30)\n",
    "# plt.title('Probability Distribution of 30th DnaA Binding Times')\n",
    "plt.ylabel('Number of Unique Ribosomal Proteins')\n",
    "plt.xlabel('Protein Count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(metIDListRiboPtn)\n",
    "print(riboPtnDist)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparing the model to the proteomics distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the proteomics count and count generated after 105 minutes for all proteins.\n",
    "\n",
    "PtnDistList = []\n",
    "PtnDist = []\n",
    "ProteomeDist = []\n",
    "\n",
    "def getProteomics(newMetID, jcvi2ID):\n",
    "    \n",
    "    # Check if protein quantification is available.\n",
    "    try:\n",
    "        if jcvi2ID.startswith(\"JCVIman_\"):\n",
    "            aoeID = manGPRPD.loc[ manGPRPD.MM == jcvi2ID.replace(\"JCVIman_\",\"\"), \"AOE\" ].values[0]\n",
    "        else:\n",
    "            aoeID = J2toAOE[ jcvi2ID ]\n",
    "        \n",
    "        ptnCount = max(0,round(proteomPD.loc[ proteomPD.Protein == aoeID ].iloc[0,21]))\n",
    "#         \n",
    "        ptnName  = proteomPD.loc[ proteomPD.Protein == aoeID ].iloc[0,1].replace(\n",
    "            \" [synthetic bacterium JCVI-Syn3.0]\",\"\")\n",
    "        \n",
    "#         ptnConcentration = ptnCount*countToMiliMol\n",
    "    except:\n",
    "        print(\"WARNING: No protein count for\", newMetID)\n",
    "        print(\"Using default protein concentration.\")\n",
    "\n",
    "        ptnName = newMetID\n",
    "        ptnCount = 0\n",
    "#         ptnConcentration = defaultPtnConcentration\n",
    "\n",
    "        ptnNoQuant.add(newMetID)\n",
    "    \n",
    "    return ptnCount, ptnName\n",
    "\n",
    "for ptnID in ModelSpecies:\n",
    "    if \"PTN_JCVI\" in ptnID:\n",
    "        PtnDistList.append(ptnID)\n",
    "        \n",
    "# for index, row in riboPtnMetDF.iterrows():\n",
    "#     result = row[\"species\"]\n",
    "#     if result:\n",
    "#         PtnDistList.append( result )\n",
    "\n",
    "# for index, row in PtnMetDF.iterrows():\n",
    "#     result = row[\"species\"]\n",
    "#     if result:\n",
    "#         PtnDistList.append( result )\n",
    "        \n",
    "for ptn in PtnDistList:\n",
    "    \n",
    "    if 'JCVISYN3A' in ptn:\n",
    "        \n",
    "        locus_code = ptn.split('3A_')[1].replace('_c','')\n",
    "        mmcode = 'MMSYN1_' + locus_code\n",
    "        \n",
    "        try:\n",
    "            jcvi2ID = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 13].strip()\n",
    "        except:\n",
    "            jcvi2ID = \"JCVIunk_\" + mmcode\n",
    "\n",
    "        ptnCount, ptnName = getProteomics(ptn, jcvi2ID)\n",
    "\n",
    "    else:\n",
    "        GeneLength = ''\n",
    "        ptnName = ''\n",
    "        ptnCount = 0\n",
    "        print('No proteomics')\n",
    "        \n",
    "#     if 'clear' in function:\n",
    "                \n",
    "#             print(highPtns,': ',ptnName)\n",
    "#             print('Function: ',function)\n",
    "#             print('Model ID: ',ptn)\n",
    "#             print('Gene Length: ',GeneLength)\n",
    "#             print('Proteomics: ',proteomics)\n",
    "#             print('Scaled Abundance after 105 min: ',scaled_abundance[-1])\n",
    "#             print()\n",
    "    \n",
    "    proteomics = ptnCount\n",
    "    \n",
    "    if proteomics > 0:\n",
    "        avgTrace = AverageSpec(ptn)\n",
    "        final_count = avgTrace[-1] - avgTrace[0]\n",
    "        \n",
    "        ptnName = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 8].strip()\n",
    "        GeneLength = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 2]\n",
    "        function = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 10].strip()\n",
    "        ProteomeDist.append(proteomics)\n",
    "        PtnDist.append(final_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(ProteomeDist),len(PtnDist))\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "plt.figure(figsize=(8,6))\n",
    "plt.hist((ProteomeDist,PtnDist),bins = 25,color = ['r','b'])\n",
    "plt.legend(('Proteomics','Proteins Generated'))\n",
    "# h1 = plt.hist(PtnDist, bins = 30, color = 'r')\n",
    "# h2 = plt.hist(ProteomeDist, bins = 30, color = 'b')\n",
    "# plt.title('Probability Distribution of 30th DnaA Binding Times')\n",
    "plt.xlabel('Protein Count')\n",
    "plt.ylabel('Number of Unique Proteins')\n",
    "# plt.xscale('log')\n",
    "# plt.xlim(1,10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the proteomics counts for all proteins with unclear function.\n",
    "\n",
    "PtnDistList = []\n",
    "# PtnDist = []\n",
    "UncProteomeDist = []\n",
    "\n",
    "def getProteomics(newMetID, jcvi2ID):\n",
    "    \n",
    "    # Check if protein quantification is available.\n",
    "    try:\n",
    "        if jcvi2ID.startswith(\"JCVIman_\"):\n",
    "            aoeID = manGPRPD.loc[ manGPRPD.MM == jcvi2ID.replace(\"JCVIman_\",\"\"), \"AOE\" ].values[0]\n",
    "        else:\n",
    "            aoeID = J2toAOE[ jcvi2ID ]\n",
    "        \n",
    "        ptnCount = max(0,round(proteomPD.loc[ proteomPD.Protein == aoeID ].iloc[0,21]))\n",
    "#         \n",
    "        ptnName  = proteomPD.loc[ proteomPD.Protein == aoeID ].iloc[0,1].replace(\n",
    "            \" [synthetic bacterium JCVI-Syn3.0]\",\"\")\n",
    "        \n",
    "#         ptnConcentration = ptnCount*countToMiliMol\n",
    "    except:\n",
    "#         print(\"WARNING: No protein count for\", newMetID)\n",
    "#         print(\"Using default protein concentration.\")\n",
    "\n",
    "        ptnName = newMetID\n",
    "        ptnCount = 0\n",
    "#         ptnConcentration = defaultPtnConcentration\n",
    "\n",
    "        ptnNoQuant.add(newMetID)\n",
    "    \n",
    "    return ptnCount, ptnName\n",
    "\n",
    "for ptnID in ModelSpecies:\n",
    "    if \"PTN_JCVI\" in ptnID:\n",
    "        PtnDistList.append(ptnID)\n",
    "        \n",
    "# for index, row in riboPtnMetDF.iterrows():\n",
    "#     result = row[\"species\"]\n",
    "#     if result:\n",
    "#         PtnDistList.append( result )\n",
    "\n",
    "# for index, row in PtnMetDF.iterrows():\n",
    "#     result = row[\"species\"]\n",
    "#     if result:\n",
    "#         PtnDistList.append( result )\n",
    "        \n",
    "for ptn in PtnDistList:\n",
    "    \n",
    "    if 'JCVISYN3A' in ptn:\n",
    "        \n",
    "        locus_code = ptn.split('3A_')[1].replace('_c','')\n",
    "        mmcode = 'MMSYN1_' + locus_code\n",
    "        \n",
    "        try:\n",
    "            jcvi2ID = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 13].strip()\n",
    "        except:\n",
    "            jcvi2ID = \"JCVIunk_\" + mmcode\n",
    "\n",
    "        ptnCount, ptnName = getProteomics(ptn, jcvi2ID)\n",
    "\n",
    "    else:\n",
    "        GeneLength = ''\n",
    "        ptnName = ''\n",
    "        ptnCount = 0\n",
    "        print('No proteomics')\n",
    "        \n",
    "#     if 'clear' in function:\n",
    "                \n",
    "#             print(highPtns,': ',ptnName)\n",
    "#             print('Function: ',function)\n",
    "#             print('Model ID: ',ptn)\n",
    "#             print('Gene Length: ',GeneLength)\n",
    "#             print('Proteomics: ',proteomics)\n",
    "#             print('Scaled Abundance after 105 min: ',scaled_abundance[-1])\n",
    "#             print()\n",
    "    \n",
    "    proteomics = ptnCount\n",
    "    \n",
    "    if proteomics > 0:\n",
    "#         avgTrace = AverageSpec(ptn)\n",
    "#         final_count = avgTrace[-1] - avgTrace[0]\n",
    "        \n",
    "        ptnName = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 8].strip()\n",
    "        GeneLength = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 2]\n",
    "        function = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 10].strip()\n",
    "        if 'clear' in function:\n",
    "            print(ptn)\n",
    "            print(ptnName)\n",
    "            print(function)\n",
    "            print(proteomics)\n",
    "            print('')\n",
    "            UncProteomeDist.append(proteomics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Histogram of proteomics for proteins with unclear function.\n",
    "\n",
    "print(len(UncProteomeDist))\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "plt.figure(figsize=(8,6))\n",
    "plt.hist(UncProteomeDist, bins = 20, color = 'b')\n",
    "plt.title('Proteomics Counts for Proteins of Unclear Function')\n",
    "plt.xlabel('Protein Count')\n",
    "plt.ylabel('Number of Unique Proteins')\n",
    "# plt.xscale('log')\n",
    "# plt.xlim(1,10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get kinetic and sequence data for ribosomal prorteins.\n",
    "\n",
    "# with open(\"riboPtnInfo.csv\", \"a\", newline='') as ltc:\n",
    "#     wr = csv.writer(ltc, dialect='excel')\n",
    "#     wr.writerow(RiboPtnNames)\n",
    "    \n",
    "# with open(\"riboPtnInfo.csv\", \"a\", newline='') as ltc:\n",
    "#     wr = csv.writer(ltc, dialect='excel')\n",
    "#     wr.writerow(RiboPtnLens)\n",
    "    \n",
    "# with open(\"riboPtnInfo.csv\", \"a\", newline='') as ltc:\n",
    "#     wr = csv.writer(ltc, dialect='excel')\n",
    "#     wr.writerow(RiboPtnTrscRates)\n",
    "    \n",
    "# with open(\"riboPtnInfo.csv\", \"a\", newline='') as ltc:\n",
    "#     wr = csv.writer(ltc, dialect='excel')\n",
    "#     wr.writerow(RiboPtnTranslatRates)\n",
    "    \n",
    "# with open(\"riboPtnInfo.csv\", \"a\", newline='') as ltc:\n",
    "#     wr = csv.writer(ltc, dialect='excel')\n",
    "#     wr.writerow(RiboPtnProteomics)\n",
    "    \n",
    "# with open(\"riboPtnInfo.csv\", \"a\", newline='') as ltc:\n",
    "#     wr = csv.writer(ltc, dialect='excel')\n",
    "#     wr.writerow(RiboPtnProd)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ATP Costs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "time = np.arange(0,simtime+1,1) # list of time steps, for x axis\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "ATP_costs = ['ATP_rep','ATP_trsc','ATP_translat','ATP_mRNAdeg','ATP_ptndeg']\n",
    "\n",
    "Total_ATP_cost = 0\n",
    "\n",
    "for cost in ATP_costs:\n",
    "    avgTrace = AverageSpec(cost)\n",
    "    plt.plot(time/60,avgTrace)\n",
    "    Total_ATP_cost = Total_ATP_cost + avgTrace[-1]\n",
    "    print(cost,': ',int(avgTrace[-1]),':',int(avgTrace[-1]*countToMiliMol),'mM')\n",
    "\n",
    "print('Total ATP Cost: ',Total_ATP_cost,':',Total_ATP_cost*countToMiliMol)    \n",
    "\n",
    "# plt.yscale('log')\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Number of ATP Hydrolyzed')\n",
    "plt.legend(ATP_costs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
